{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get a set of enhancers known to be constrained or not, together with their Chen scores "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "CONSTRAINT_TOOLS = '/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools'\n",
    "CONSTRAINT_TOOLS_DATA = '/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>negative_new_chen_score_window</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21346107</td>\n",
       "      <td>21348107</td>\n",
       "      <td>-3.432213</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93909844</td>\n",
       "      <td>93911044</td>\n",
       "      <td>-2.645724</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155293209</td>\n",
       "      <td>155293609</td>\n",
       "      <td>-4.190955</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>160032410</td>\n",
       "      <td>160033810</td>\n",
       "      <td>-2.719339</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>160032410</td>\n",
       "      <td>160033810</td>\n",
       "      <td>-0.167476</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>chr20</td>\n",
       "      <td>46114761</td>\n",
       "      <td>46114961</td>\n",
       "      <td>0.241142</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>chr22</td>\n",
       "      <td>28787612</td>\n",
       "      <td>28789612</td>\n",
       "      <td>2.138137</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>chr22</td>\n",
       "      <td>28787612</td>\n",
       "      <td>28789612</td>\n",
       "      <td>0.488883</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>chr22</td>\n",
       "      <td>28787612</td>\n",
       "      <td>28789612</td>\n",
       "      <td>1.762075</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>chr22</td>\n",
       "      <td>38020193</td>\n",
       "      <td>38021993</td>\n",
       "      <td>-3.399608</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end  negative_new_chen_score_window  constrained  \\\n",
       "0     chr1   21346107   21348107                       -3.432213         True   \n",
       "1     chr1   93909844   93911044                       -2.645724         True   \n",
       "2     chr1  155293209  155293609                       -4.190955         True   \n",
       "3     chr1  160032410  160033810                       -2.719339         True   \n",
       "4     chr1  160032410  160033810                       -0.167476         True   \n",
       "..     ...        ...        ...                             ...          ...   \n",
       "116  chr20   46114761   46114961                        0.241142         True   \n",
       "117  chr22   28787612   28789612                        2.138137         True   \n",
       "118  chr22   28787612   28789612                        0.488883         True   \n",
       "119  chr22   28787612   28789612                        1.762075         True   \n",
       "120  chr22   38020193   38021993                       -3.399608         True   \n",
       "\n",
       "                   tag  \n",
       "0    disease_enhancers  \n",
       "1    disease_enhancers  \n",
       "2    disease_enhancers  \n",
       "3    disease_enhancers  \n",
       "4    disease_enhancers  \n",
       "..                 ...  \n",
       "116  disease_enhancers  \n",
       "117  disease_enhancers  \n",
       "118  disease_enhancers  \n",
       "119  disease_enhancers  \n",
       "120  disease_enhancers  \n",
       "\n",
       "[121 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def read_disease_enhancers_intersect_chen_windows():\n",
    "  # download-process-data/khurana/README.md\n",
    "  df = pd.read_csv(\n",
    "      f\"{CONSTRAINT_TOOLS_DATA}/khurana/disease-enhancers-intersect-chen-windows.bed\", \n",
    "      sep = '\\t',\n",
    "      names = [\n",
    "          'chrom_enhancer', 'start_enhancer', 'end_enhancer', \n",
    "          'disease', 'enhancer_predicted_LoF_tolerance_prob', 'enhancer_predicted_LoF_tolerance_status', \n",
    "          'chrom_window', 'start_window', 'end_window', \n",
    "          'new_chen_score_window'\n",
    "      ]\n",
    "  )\n",
    "  df = df.drop_duplicates() \n",
    "  df['negative_new_chen_score_window'] = -df['new_chen_score_window']\n",
    "  df = df.drop(columns=[\n",
    "    'new_chen_score_window',\n",
    "    'chrom_window', 'start_window', 'end_window',\n",
    "    'disease', 'enhancer_predicted_LoF_tolerance_prob', 'enhancer_predicted_LoF_tolerance_status'\n",
    "  ])\n",
    "  df = df.rename(columns={\n",
    "    'chrom_enhancer': 'chrom', \n",
    "    'start_enhancer': 'start', \n",
    "    'end_enhancer': 'end'\n",
    "  })\n",
    "  df['constrained'] = True\n",
    "  df['tag'] = 'disease_enhancers'\n",
    "  return df \n",
    "\n",
    "read_disease_enhancers_intersect_chen_windows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>negative_new_chen_score_window</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>5785468</td>\n",
       "      <td>5785668</td>\n",
       "      <td>2.210384</td>\n",
       "      <td>False</td>\n",
       "      <td>pacbio_deleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr3</td>\n",
       "      <td>121543753</td>\n",
       "      <td>121543953</td>\n",
       "      <td>-1.011635</td>\n",
       "      <td>False</td>\n",
       "      <td>pacbio_deleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>69152682</td>\n",
       "      <td>69153682</td>\n",
       "      <td>2.649651</td>\n",
       "      <td>False</td>\n",
       "      <td>pacbio_deleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>69152682</td>\n",
       "      <td>69153682</td>\n",
       "      <td>1.650749</td>\n",
       "      <td>False</td>\n",
       "      <td>pacbio_deleted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>6969081</td>\n",
       "      <td>6970081</td>\n",
       "      <td>-2.492414</td>\n",
       "      <td>False</td>\n",
       "      <td>pacbio_deleted</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start        end  negative_new_chen_score_window  constrained  \\\n",
       "0   chr2    5785468    5785668                        2.210384        False   \n",
       "1   chr3  121543753  121543953                       -1.011635        False   \n",
       "2   chr4   69152682   69153682                        2.649651        False   \n",
       "3   chr4   69152682   69153682                        1.650749        False   \n",
       "4  chr17    6969081    6970081                       -2.492414        False   \n",
       "\n",
       "              tag  \n",
       "0  pacbio_deleted  \n",
       "1  pacbio_deleted  \n",
       "2  pacbio_deleted  \n",
       "3  pacbio_deleted  \n",
       "4  pacbio_deleted  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def read_pacbio_deleted_enhancers_intersect_chen_windows():\n",
    "  # download-process-data/khurana/README.md\n",
    "  df = pd.read_csv(\n",
    "      f\"{CONSTRAINT_TOOLS_DATA}/khurana/pacbio-deleted-enhancers-intersect-chen-windows.bed\", \n",
    "      sep = '\\t',\n",
    "      names = [\n",
    "          'chrom_enhancer', 'start_enhancer', 'end_enhancer', \n",
    "          'family', 'enhancer_predicted_LoF_tolerance_prob', 'enhancer_predicted_LoF_tolerance_status', \n",
    "          'chrom_window', 'start_window', 'end_window', \n",
    "          'new_chen_score_window'\n",
    "      ]\n",
    "  )\n",
    "  df = df.drop_duplicates() \n",
    "  df['negative_new_chen_score_window'] = -df['new_chen_score_window']\n",
    "  df = df.drop(columns=[\n",
    "    'new_chen_score_window',\n",
    "    'chrom_window', 'start_window', 'end_window',\n",
    "    'family', 'enhancer_predicted_LoF_tolerance_prob', 'enhancer_predicted_LoF_tolerance_status'\n",
    "  ])\n",
    "  df = df.rename(columns={\n",
    "    'chrom_enhancer': 'chrom', \n",
    "    'start_enhancer': 'start', \n",
    "    'end_enhancer': 'end'\n",
    "  })\n",
    "  df['constrained'] = False\n",
    "  df['tag'] = 'pacbio_deleted'\n",
    "  return df \n",
    "\n",
    "read_pacbio_deleted_enhancers_intersect_chen_windows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>negative_new_chen_score_window</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>831220</td>\n",
       "      <td>832820</td>\n",
       "      <td>-1.248459</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1508220</td>\n",
       "      <td>1509820</td>\n",
       "      <td>-1.731383</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1508220</td>\n",
       "      <td>1509820</td>\n",
       "      <td>-1.670577</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8124140</td>\n",
       "      <td>8124940</td>\n",
       "      <td>-0.841769</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12684990</td>\n",
       "      <td>12685190</td>\n",
       "      <td>-1.311703</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>chr22</td>\n",
       "      <td>36177152</td>\n",
       "      <td>36177552</td>\n",
       "      <td>-1.206580</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835</th>\n",
       "      <td>chr22</td>\n",
       "      <td>36548353</td>\n",
       "      <td>36550153</td>\n",
       "      <td>-0.796912</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <td>chr22</td>\n",
       "      <td>44535320</td>\n",
       "      <td>44535920</td>\n",
       "      <td>-0.025900</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>837</th>\n",
       "      <td>chr22</td>\n",
       "      <td>48390588</td>\n",
       "      <td>48391588</td>\n",
       "      <td>-0.083336</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838</th>\n",
       "      <td>chr22</td>\n",
       "      <td>48390588</td>\n",
       "      <td>48391588</td>\n",
       "      <td>-1.782994</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>839 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  negative_new_chen_score_window  constrained  \\\n",
       "0     chr1    831220    832820                       -1.248459        False   \n",
       "1     chr1   1508220   1509820                       -1.731383        False   \n",
       "2     chr1   1508220   1509820                       -1.670577        False   \n",
       "3     chr1   8124140   8124940                       -0.841769        False   \n",
       "4     chr1  12684990  12685190                       -1.311703        False   \n",
       "..     ...       ...       ...                             ...          ...   \n",
       "834  chr22  36177152  36177552                       -1.206580        False   \n",
       "835  chr22  36548353  36550153                       -0.796912        False   \n",
       "836  chr22  44535320  44535920                       -0.025900        False   \n",
       "837  chr22  48390588  48391588                       -0.083336        False   \n",
       "838  chr22  48390588  48391588                       -1.782994        False   \n",
       "\n",
       "              tag  \n",
       "0    lof_tolerant  \n",
       "1    lof_tolerant  \n",
       "2    lof_tolerant  \n",
       "3    lof_tolerant  \n",
       "4    lof_tolerant  \n",
       "..            ...  \n",
       "834  lof_tolerant  \n",
       "835  lof_tolerant  \n",
       "836  lof_tolerant  \n",
       "837  lof_tolerant  \n",
       "838  lof_tolerant  \n",
       "\n",
       "[839 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get enhancers that are homozygously deleted in 1000 Genomes Project\n",
    "def read_lof_tolerant_enhancers_intersect_chen_windows():\n",
    "  # download-process-data/khurana/README.md\n",
    "  df = pd.read_csv(\n",
    "      f\"{CONSTRAINT_TOOLS_DATA}/khurana/lof-tolerant-enhancers-intersect-chen-windows.bed\", \n",
    "      sep = '\\t',\n",
    "      names = [\n",
    "          'chrom_enhancer_hg38', 'start_enhancer_hg38', 'end_enhancer_hg38', \n",
    "          'enhancer_hg19', 'unknown',\n",
    "          'chrom_window', 'start_window', 'end_window', \n",
    "          'new_chen_score_window'\n",
    "      ]\n",
    "  )\n",
    "  df = df.drop_duplicates() \n",
    "  df['negative_new_chen_score_window'] = -df['new_chen_score_window']\n",
    "  df = df.drop(columns=[\n",
    "    'new_chen_score_window',\n",
    "    'chrom_window', 'start_window', 'end_window',\n",
    "    'enhancer_hg19', 'unknown'\n",
    "  ])\n",
    "  df = df.rename(columns={\n",
    "    'chrom_enhancer_hg38': 'chrom', \n",
    "    'start_enhancer_hg38': 'start', \n",
    "    'end_enhancer_hg38': 'end'\n",
    "  })\n",
    "  df['constrained'] = False\n",
    "  df['tag'] = 'lof_tolerant'\n",
    "  return df \n",
    "\n",
    "read_lof_tolerant_enhancers_intersect_chen_windows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>negative_new_chen_score_window</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61587728</td>\n",
       "      <td>61589928</td>\n",
       "      <td>-0.884984</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61587728</td>\n",
       "      <td>61589928</td>\n",
       "      <td>-3.667483</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61587728</td>\n",
       "      <td>61589928</td>\n",
       "      <td>-1.787719</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>87355917</td>\n",
       "      <td>87357117</td>\n",
       "      <td>-1.174601</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>87355917</td>\n",
       "      <td>87357117</td>\n",
       "      <td>-2.971624</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>87355917</td>\n",
       "      <td>87357117</td>\n",
       "      <td>-1.597840</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169941659</td>\n",
       "      <td>169943659</td>\n",
       "      <td>-0.732262</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169941659</td>\n",
       "      <td>169943659</td>\n",
       "      <td>-0.494991</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>66069666</td>\n",
       "      <td>66071266</td>\n",
       "      <td>-1.628703</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr2</td>\n",
       "      <td>66069666</td>\n",
       "      <td>66071266</td>\n",
       "      <td>-1.923346</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start        end  negative_new_chen_score_window  constrained  \\\n",
       "0   chr1   61587728   61589928                       -0.884984         True   \n",
       "1   chr1   61587728   61589928                       -3.667483         True   \n",
       "2   chr1   61587728   61589928                       -1.787719         True   \n",
       "3   chr1   87355917   87357117                       -1.174601         True   \n",
       "4   chr1   87355917   87357117                       -2.971624         True   \n",
       "5   chr1   87355917   87357117                       -1.597840         True   \n",
       "9   chr1  169941659  169943659                       -0.732262         True   \n",
       "10  chr1  169941659  169943659                       -0.494991         True   \n",
       "11  chr2   66069666   66071266                       -1.628703         True   \n",
       "12  chr2   66069666   66071266                       -1.923346         True   \n",
       "\n",
       "                  tag  \n",
       "0   low_lof_tolerance  \n",
       "1   low_lof_tolerance  \n",
       "2   low_lof_tolerance  \n",
       "3   low_lof_tolerance  \n",
       "4   low_lof_tolerance  \n",
       "5   low_lof_tolerance  \n",
       "9   low_lof_tolerance  \n",
       "10  low_lof_tolerance  \n",
       "11  low_lof_tolerance  \n",
       "12  low_lof_tolerance  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get low-lof-tolerance enhancers\n",
    "def read_low_lof_tolerance_enhancers_intersect_chen_windows():\n",
    "  # download-process-data/khurana/README.md\n",
    "  df = pd.read_csv(\n",
    "      f\"{CONSTRAINT_TOOLS_DATA}/khurana/low-lof-tolerance-enhancers-intersect-chen-windows.bed\", \n",
    "      sep = '\\t',\n",
    "      names = [\n",
    "          'chrom_enhancer_hg38', 'start_enhancer_hg38', 'end_enhancer_hg38', \n",
    "          'enhancer_hg19', 'unknown',\n",
    "          'chrom_window', 'start_window', 'end_window', \n",
    "          'new_chen_score_window'\n",
    "      ]\n",
    "  )\n",
    "  df = df.drop_duplicates() \n",
    "  df['negative_new_chen_score_window'] = -df['new_chen_score_window']\n",
    "  df = df.drop(columns=[\n",
    "    'new_chen_score_window',\n",
    "    'chrom_window', 'start_window', 'end_window',\n",
    "    'enhancer_hg19', 'unknown'\n",
    "  ])\n",
    "  df = df.rename(columns={\n",
    "    'chrom_enhancer_hg38': 'chrom', \n",
    "    'start_enhancer_hg38': 'start', \n",
    "    'end_enhancer_hg38': 'end'\n",
    "  })\n",
    "  df['constrained'] = True\n",
    "  df['tag'] = 'low_lof_tolerance'\n",
    "  return df \n",
    "\n",
    "read_low_lof_tolerance_enhancers_intersect_chen_windows().iloc[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "      <th>min negative_new_chen_score_window</th>\n",
       "      <th>count negative_new_chen_score_window</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>831220</td>\n",
       "      <td>832820</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.248459</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1508220</td>\n",
       "      <td>1509820</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.731383</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8124140</td>\n",
       "      <td>8124940</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-0.841769</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12684990</td>\n",
       "      <td>12685190</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.403148</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16104305</td>\n",
       "      <td>16105505</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>0.845290</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>chr9</td>\n",
       "      <td>126908121</td>\n",
       "      <td>126908321</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.462233</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129765721</td>\n",
       "      <td>129766721</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.356569</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129766921</td>\n",
       "      <td>129767321</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.356569</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>chr9</td>\n",
       "      <td>132748213</td>\n",
       "      <td>132748813</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-2.111835</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137289148</td>\n",
       "      <td>137290348</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-2.162811</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>510 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end  constrained           tag  \\\n",
       "0    chr1     831220     832820        False  lof_tolerant   \n",
       "1    chr1    1508220    1509820        False  lof_tolerant   \n",
       "2    chr1    8124140    8124940        False  lof_tolerant   \n",
       "3    chr1   12684990   12685190        False  lof_tolerant   \n",
       "4    chr1   16104305   16105505        False  lof_tolerant   \n",
       "..    ...        ...        ...          ...           ...   \n",
       "505  chr9  126908121  126908321        False  lof_tolerant   \n",
       "506  chr9  129765721  129766721        False  lof_tolerant   \n",
       "507  chr9  129766921  129767321        False  lof_tolerant   \n",
       "508  chr9  132748213  132748813        False  lof_tolerant   \n",
       "509  chr9  137289148  137290348        False  lof_tolerant   \n",
       "\n",
       "     min negative_new_chen_score_window  count negative_new_chen_score_window  \n",
       "0                             -1.248459                                     1  \n",
       "1                             -1.731383                                     2  \n",
       "2                             -0.841769                                     1  \n",
       "3                             -1.403148                                     2  \n",
       "4                              0.845290                                     1  \n",
       "..                                  ...                                   ...  \n",
       "505                           -1.462233                                     1  \n",
       "506                           -1.356569                                     1  \n",
       "507                           -1.356569                                     1  \n",
       "508                           -2.111835                                     1  \n",
       "509                           -2.162811                                     1  \n",
       "\n",
       "[510 rows x 7 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def aggregate_over_windows(df): \n",
    "    columns = df.columns.tolist()\n",
    "    group_columns = [column for column in columns if column != 'negative_new_chen_score_window']\n",
    "    groups = df.groupby(group_columns)\n",
    "    aggregation_functions = {'negative_new_chen_score_window': ['min', 'count']}\n",
    "    aggregated = groups.agg(aggregation_functions)  \n",
    "    df = aggregated.reset_index()\n",
    "    df.columns = [' '.join(col[::-1]).strip() for col in df.columns.values]\n",
    "    return df\n",
    "\n",
    "aggregate_over_windows(read_lof_tolerant_enhancers_intersect_chen_windows())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of enhancers that are constrained (True) or not (False):\n",
      "False    514\n",
      "True     118\n",
      "Name: constrained, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "      <th>min negative_new_chen_score_window</th>\n",
       "      <th>count negative_new_chen_score_window</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21346107</td>\n",
       "      <td>21348107</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "      <td>-3.432213</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93909844</td>\n",
       "      <td>93911044</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "      <td>-2.645724</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155293209</td>\n",
       "      <td>155293609</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "      <td>-4.190955</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>160032410</td>\n",
       "      <td>160033810</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "      <td>-2.719339</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>173915262</td>\n",
       "      <td>173915462</td>\n",
       "      <td>True</td>\n",
       "      <td>disease_enhancers</td>\n",
       "      <td>-1.790974</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>chr7</td>\n",
       "      <td>31361986</td>\n",
       "      <td>31364186</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "      <td>-1.247859</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>chr7</td>\n",
       "      <td>54568507</td>\n",
       "      <td>54569507</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "      <td>-1.270480</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>chr7</td>\n",
       "      <td>97011888</td>\n",
       "      <td>97013488</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "      <td>-0.365397</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>chr8</td>\n",
       "      <td>98415972</td>\n",
       "      <td>98417172</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "      <td>0.620860</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3400800</td>\n",
       "      <td>3402600</td>\n",
       "      <td>True</td>\n",
       "      <td>low_lof_tolerance</td>\n",
       "      <td>-0.622948</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>632 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end  constrained                tag  \\\n",
       "0    chr1   21346107   21348107         True  disease_enhancers   \n",
       "1    chr1   93909844   93911044         True  disease_enhancers   \n",
       "2    chr1  155293209  155293609         True  disease_enhancers   \n",
       "3    chr1  160032410  160033810         True  disease_enhancers   \n",
       "4    chr1  173915262  173915462         True  disease_enhancers   \n",
       "..    ...        ...        ...          ...                ...   \n",
       "627  chr7   31361986   31364186         True  low_lof_tolerance   \n",
       "628  chr7   54568507   54569507         True  low_lof_tolerance   \n",
       "629  chr7   97011888   97013488         True  low_lof_tolerance   \n",
       "630  chr8   98415972   98417172         True  low_lof_tolerance   \n",
       "631  chr9    3400800    3402600         True  low_lof_tolerance   \n",
       "\n",
       "     min negative_new_chen_score_window  count negative_new_chen_score_window  \n",
       "0                             -3.432213                                     1  \n",
       "1                             -2.645724                                     1  \n",
       "2                             -4.190955                                     1  \n",
       "3                             -2.719339                                     2  \n",
       "4                             -1.790974                                     1  \n",
       "..                                  ...                                   ...  \n",
       "627                           -1.247859                                     4  \n",
       "628                           -1.270480                                     2  \n",
       "629                           -0.365397                                     2  \n",
       "630                            0.620860                                     1  \n",
       "631                           -0.622948                                     3  \n",
       "\n",
       "[632 rows x 7 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_labeled_enhancers(): \n",
    "  df = pd.concat([\n",
    "    aggregate_over_windows(read_disease_enhancers_intersect_chen_windows()),\n",
    "    aggregate_over_windows(read_pacbio_deleted_enhancers_intersect_chen_windows()),\n",
    "    aggregate_over_windows(read_lof_tolerant_enhancers_intersect_chen_windows()),\n",
    "    aggregate_over_windows(read_low_lof_tolerance_enhancers_intersect_chen_windows())  \n",
    "  ])\n",
    "  df = df.reset_index(drop=True)\n",
    "  print('number of enhancers that are constrained (True) or not (False):') \n",
    "  print(df['constrained'].value_counts())\n",
    "  df.to_csv(f\"{CONSTRAINT_TOOLS_DATA}/khurana/labeled-enhancers.bed\", sep='\\t', index=False)\n",
    "  return df \n",
    "\n",
    "pd.set_option('display.max_rows', 10)\n",
    "\n",
    "create_labeled_enhancers()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intersect Khurana enhancers with homozygous deletions from TopMed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash /scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools/experiments/germline-model/chen-et-al-2022/intersect-labeled-enhancers-with-topmed-homozygous-deletions.sh\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_enhancer</th>\n",
       "      <th>start_enhancer</th>\n",
       "      <th>end_enhancer</th>\n",
       "      <th>constrained</th>\n",
       "      <th>tag</th>\n",
       "      <th>min negative_new_chen_score_window</th>\n",
       "      <th>count negative_new_chen_score_window</th>\n",
       "      <th>custom_max SVLEN</th>\n",
       "      <th>collapse SVTYPE</th>\n",
       "      <th>count SVTYPE</th>\n",
       "      <th>custom_list SV_ID</th>\n",
       "      <th>custom_max HomAlt</th>\n",
       "      <th>custom_max enhancer-deletion-overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>831220</td>\n",
       "      <td>832820</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.248459</td>\n",
       "      <td>1</td>\n",
       "      <td>2521</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1</td>\n",
       "      <td>[DEL_1:831216-833736]</td>\n",
       "      <td>99</td>\n",
       "      <td>1600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1508220</td>\n",
       "      <td>1509820</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.731383</td>\n",
       "      <td>2</td>\n",
       "      <td>74900</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3</td>\n",
       "      <td>[DEL_1:1508186-1510388, DEL_1:1443801-1518700,...</td>\n",
       "      <td>262</td>\n",
       "      <td>1600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8124140</td>\n",
       "      <td>8124940</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-0.841769</td>\n",
       "      <td>1</td>\n",
       "      <td>9600</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>[DEL_1:8121301-8130900, DEL_1:8122347-8130816]</td>\n",
       "      <td>65</td>\n",
       "      <td>800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12684990</td>\n",
       "      <td>12685190</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.403148</td>\n",
       "      <td>2</td>\n",
       "      <td>3222</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1</td>\n",
       "      <td>[DEL_1:12683396-12686617]</td>\n",
       "      <td>175</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16104305</td>\n",
       "      <td>16105505</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>0.845290</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>chr9</td>\n",
       "      <td>126908121</td>\n",
       "      <td>126908321</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.462233</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129765721</td>\n",
       "      <td>129766721</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.356569</td>\n",
       "      <td>1</td>\n",
       "      <td>5100</td>\n",
       "      <td>DEL</td>\n",
       "      <td>4</td>\n",
       "      <td>[DEL_9:129763101-129768200, DEL_9:129764101-12...</td>\n",
       "      <td>116</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129766921</td>\n",
       "      <td>129767321</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-1.356569</td>\n",
       "      <td>1</td>\n",
       "      <td>5100</td>\n",
       "      <td>DEL</td>\n",
       "      <td>4</td>\n",
       "      <td>[DEL_9:129763101-129768200, DEL_9:129764101-12...</td>\n",
       "      <td>116</td>\n",
       "      <td>400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>chr9</td>\n",
       "      <td>132748213</td>\n",
       "      <td>132748813</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-2.111835</td>\n",
       "      <td>1</td>\n",
       "      <td>4608</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1</td>\n",
       "      <td>[DEL_9:132747349-132751956]</td>\n",
       "      <td>36</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137289148</td>\n",
       "      <td>137290348</td>\n",
       "      <td>False</td>\n",
       "      <td>lof_tolerant</td>\n",
       "      <td>-2.162811</td>\n",
       "      <td>1</td>\n",
       "      <td>7300</td>\n",
       "      <td>DEL</td>\n",
       "      <td>5</td>\n",
       "      <td>[DEL_9:137284801-137291800, DEL_9:137285593-13...</td>\n",
       "      <td>143</td>\n",
       "      <td>1200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>632 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom_enhancer  start_enhancer  end_enhancer  constrained           tag  \\\n",
       "0             chr1          831220        832820        False  lof_tolerant   \n",
       "1             chr1         1508220       1509820        False  lof_tolerant   \n",
       "2             chr1         8124140       8124940        False  lof_tolerant   \n",
       "3             chr1        12684990      12685190        False  lof_tolerant   \n",
       "4             chr1        16104305      16105505        False  lof_tolerant   \n",
       "..             ...             ...           ...          ...           ...   \n",
       "627           chr9       126908121     126908321        False  lof_tolerant   \n",
       "628           chr9       129765721     129766721        False  lof_tolerant   \n",
       "629           chr9       129766921     129767321        False  lof_tolerant   \n",
       "630           chr9       132748213     132748813        False  lof_tolerant   \n",
       "631           chr9       137289148     137290348        False  lof_tolerant   \n",
       "\n",
       "     min negative_new_chen_score_window  count negative_new_chen_score_window  \\\n",
       "0                             -1.248459                                     1   \n",
       "1                             -1.731383                                     2   \n",
       "2                             -0.841769                                     1   \n",
       "3                             -1.403148                                     2   \n",
       "4                              0.845290                                     1   \n",
       "..                                  ...                                   ...   \n",
       "627                           -1.462233                                     1   \n",
       "628                           -1.356569                                     1   \n",
       "629                           -1.356569                                     1   \n",
       "630                           -2.111835                                     1   \n",
       "631                           -2.162811                                     1   \n",
       "\n",
       "    custom_max SVLEN collapse SVTYPE count SVTYPE  \\\n",
       "0               2521             DEL            1   \n",
       "1              74900             DEL            3   \n",
       "2               9600             DEL            2   \n",
       "3               3222             DEL            1   \n",
       "4                  .               .            .   \n",
       "..               ...             ...          ...   \n",
       "627                .               .            .   \n",
       "628             5100             DEL            4   \n",
       "629             5100             DEL            4   \n",
       "630             4608             DEL            1   \n",
       "631             7300             DEL            5   \n",
       "\n",
       "                                     custom_list SV_ID custom_max HomAlt  \\\n",
       "0                                [DEL_1:831216-833736]                99   \n",
       "1    [DEL_1:1508186-1510388, DEL_1:1443801-1518700,...               262   \n",
       "2       [DEL_1:8121301-8130900, DEL_1:8122347-8130816]                65   \n",
       "3                            [DEL_1:12683396-12686617]               175   \n",
       "4                                                    .                 .   \n",
       "..                                                 ...               ...   \n",
       "627                                                  .                 .   \n",
       "628  [DEL_9:129763101-129768200, DEL_9:129764101-12...               116   \n",
       "629  [DEL_9:129763101-129768200, DEL_9:129764101-12...               116   \n",
       "630                        [DEL_9:132747349-132751956]                36   \n",
       "631  [DEL_9:137284801-137291800, DEL_9:137285593-13...               143   \n",
       "\n",
       "     custom_max enhancer-deletion-overlap  \n",
       "0                                    1600  \n",
       "1                                    1600  \n",
       "2                                     800  \n",
       "3                                     200  \n",
       "4                                       0  \n",
       "..                                    ...  \n",
       "627                                     0  \n",
       "628                                  1000  \n",
       "629                                   400  \n",
       "630                                   600  \n",
       "631                                  1200  \n",
       "\n",
       "[632 rows x 13 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "import numpy as np \n",
    "\n",
    "def shell(cmd):\n",
    "  # run shell command and return output as a string\n",
    "  return subprocess.run(\n",
    "    cmd,\n",
    "    shell=True,\n",
    "    stdout=subprocess.PIPE,\n",
    "    stderr=subprocess.PIPE,\n",
    "  ).stdout.decode(\"utf-8\").strip()\n",
    "\n",
    "def collapse(ser): \n",
    "  unique_value, = set(ser)\n",
    "  return unique_value\n",
    "\n",
    "def count(ser): \n",
    "  if len(ser) == 1 and ser.iloc[0] == '.': return '.'\n",
    "  return len(ser) \n",
    "\n",
    "def custom_max(ser): \n",
    "  if len(ser) == 1 and ser.iloc[0] == '.': return '.'\n",
    "  return np.max([int(value) for value in list(ser)])\n",
    "\n",
    "def custom_list(ser): \n",
    "  if len(ser) == 1 and ser.iloc[0] == '.': return '.'\n",
    "  return list(ser)\n",
    "\n",
    "def aggregate_over_deletions(df, group_columns, aggregation_functions): \n",
    "  groups = df.groupby(group_columns)\n",
    "  aggregated = groups.agg(aggregation_functions)  \n",
    "  df = aggregated.reset_index()\n",
    "  df.columns = [' '.join(col[::-1]).strip() for col in df.columns.values]\n",
    "  return df\n",
    "\n",
    "def read_enhancers_intersect_homozygous_deletions():\n",
    "  cmd = f'bash {CONSTRAINT_TOOLS}/experiments/germline-model/chen-et-al-2022/intersect-labeled-enhancers-with-topmed-homozygous-deletions.sh'\n",
    "  print(cmd)\n",
    "  print(shell(cmd))\n",
    "  df = pd.read_csv(\n",
    "    f\"{CONSTRAINT_TOOLS_DATA}/khurana/labeled-enhancers-intersect-topmed-homozygous-deletions.bed\", \n",
    "    sep = '\\t',\n",
    "  )  \n",
    "  group_columns = [\n",
    "    'chrom',\n",
    "    'start',\n",
    "    'end',\n",
    "    'constrained',\n",
    "    'tag',\n",
    "    'min negative_new_chen_score_window',\n",
    "    'count negative_new_chen_score_window',\n",
    "  ]\n",
    "  aggregation_functions = {\n",
    "    'SVLEN': [custom_max],\n",
    "    'SVTYPE': [collapse, count],\n",
    "    'SV_ID': [custom_list],\n",
    "    'HomAlt': [custom_max],\n",
    "    'enhancer-deletion-overlap': [custom_max]\n",
    "  }\n",
    "  aggregation_columns = list(aggregation_functions.keys())\n",
    "  new_columns = group_columns + aggregation_columns\n",
    "  df = df[new_columns]\n",
    "  df = aggregate_over_deletions(df, group_columns, aggregation_functions)\n",
    "  df = df.rename(columns={\n",
    "    'chrom': 'chrom_enhancer',\n",
    "    'start': 'start_enhancer',\n",
    "    'end': 'end_enhancer',\n",
    "  })\n",
    "  return df   \n",
    "  # TODO: \n",
    "  # add a column indicating whether enhancer overlaps a homozygous deletion\n",
    "  \n",
    "read_enhancers_intersect_homozygous_deletions()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using Chen zscore to predict whether an enhancer is critical or not, and improving that prediction using homozygous deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 2. for each score threshold\n",
    "#   i. predict class (constrained vs non-constrained), and compute contingency table\n",
    "#   ii. create a new contingency table in which examples predicted to be constrained by Chen, \n",
    "#       but which overlap deletions, are now predicted to be unconstrained, \n",
    "# 3. repeat for all thresholds, and measure AUROC or AUPRC, both for original contingency tables and altered ones \n",
    "\n",
    "# Under the assumption that deletions can identify false-positive calls of constraint, one would expect performance to improve"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Avoiding data leakage by defining random enhancers to be unconstrained "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using Chen zscore to predict whether a SNV is deleterious or not, and improving that prediction using homozygous deletions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 0. resources: \n",
    "#    experiments/germline-model/chen-et-al-2022/pathogenic_variant_enrichment*.ipynb\n",
    "#    experiments/germline-model/chen-et-al-2022/*clin*\n",
    "#    download-process-data/noncoding-variants-associated-with-Mendelian-traits\n",
    "#    download-process-data/compute-trustworthy-exon-adjacent-windows.sh\n",
    "#    download-process-data/clinvar-noncoding-with-negative-controls\n",
    "# 1. intersect clinvar variants with chen windows (with chen scores), \n",
    "#    instead of the other way around (as I previously did when testing for enrichment of clinvar variants in constrained windows)\n",
    "# 2. this gives a single chen score to each clinvar variant \n",
    "# 3. do the same for a set of gnomad variants of equal number \n",
    "# 4. we now have a positive and negative set of SNVs, which we can use to draw a ROC curve, by thresholding the chen score\n",
    "# 5. next, as above, we can declare those variants that overlap homozygous deletions as false positives, and recompute the ROC curve"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "constraint-tools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
