{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "CONSTRAINT_TOOLS = '/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools'\n",
    "CONSTRAINT_TOOLS_DATA = '/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools'\n",
    "\n",
    "import sys\n",
    "sys.path.append(f'{CONSTRAINT_TOOLS}/utilities')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign scores to non-exonic Chen windows "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>window overlaps enhancer</th>\n",
       "      <th>window overlaps merged_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1432000</td>\n",
       "      <td>1433000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1435000</td>\n",
       "      <td>1436000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1449000</td>\n",
       "      <td>1450000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1450000</td>\n",
       "      <td>1451000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1451000</td>\n",
       "      <td>1452000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786072</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137269000</td>\n",
       "      <td>137270000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786073</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137275000</td>\n",
       "      <td>137276000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786074</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137282000</td>\n",
       "      <td>137283000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786075</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137290000</td>\n",
       "      <td>137291000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786076</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137294000</td>\n",
       "      <td>137295000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1786077 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end  window overlaps enhancer  \\\n",
       "0        chr1    1432000    1433000                      True   \n",
       "1        chr1    1435000    1436000                      True   \n",
       "2        chr1    1449000    1450000                     False   \n",
       "3        chr1    1450000    1451000                     False   \n",
       "4        chr1    1451000    1452000                     False   \n",
       "...       ...        ...        ...                       ...   \n",
       "1786072  chr9  137269000  137270000                      True   \n",
       "1786073  chr9  137275000  137276000                      True   \n",
       "1786074  chr9  137282000  137283000                      True   \n",
       "1786075  chr9  137290000  137291000                     False   \n",
       "1786076  chr9  137294000  137295000                      True   \n",
       "\n",
       "         window overlaps merged_exon  \n",
       "0                              False  \n",
       "1                               True  \n",
       "2                               True  \n",
       "3                               True  \n",
       "4                              False  \n",
       "...                              ...  \n",
       "1786072                        False  \n",
       "1786073                        False  \n",
       "1786074                         True  \n",
       "1786075                        False  \n",
       "1786076                         True  \n",
       "\n",
       "[1786077 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from shell import shell \n",
    "from colorize import print_string_as_info\n",
    "from aggregate import aggregate\n",
    "\n",
    "def get_chen_windows_with_enhancers_exons():\n",
    "  # the following bed file was produced by experiments/germline-model/chen-et-al-2022/Fig_2a.Nonly.noisy.ipynb\n",
    "  filename = f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/mchale.kmerSizes.trainSets.noisy.enhancer-exon.bed'\n",
    "  df = pd.read_csv(filename, sep='\\t')\n",
    "  df = df[['chromosome', 'start', 'end', 'window overlaps enhancer', 'window overlaps merged_exon']]\n",
    "  df = df.rename(columns={\n",
    "    'chromosome': 'chrom'\n",
    "  })\n",
    "  return df\n",
    "\n",
    "WINDOWS_0 = get_chen_windows_with_enhancers_exons()\n",
    "WINDOWS_0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.intersect.B-map.bed\u001b[0m\n",
      "<ipython-input-6-81b67884980b>:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[col] = df[col].astype(dtype)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.intersect.paternal_recombination_rate-map.bed\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.intersect.maternal_recombination_rate-map.bed\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.maternal_recombination_rate.gBGC-tract-counts.bed\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gnocchi</th>\n",
       "      <th>N_expected</th>\n",
       "      <th>N_observed</th>\n",
       "      <th>window overlaps enhancer</th>\n",
       "      <th>window overlaps merged_exon</th>\n",
       "      <th>B</th>\n",
       "      <th>paternal_recombination_rate</th>\n",
       "      <th>maternal_recombination_rate</th>\n",
       "      <th>gBGC-tract_overlap_count</th>\n",
       "      <th>log10_paternal_recombination_rate</th>\n",
       "      <th>log10_maternal_recombination_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1432000</td>\n",
       "      <td>1433000</td>\n",
       "      <td>4.299894</td>\n",
       "      <td>338.059552</td>\n",
       "      <td>259</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.653</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1451000</td>\n",
       "      <td>1452000</td>\n",
       "      <td>0.666316</td>\n",
       "      <td>302.590657</td>\n",
       "      <td>291</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.652</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1453000</td>\n",
       "      <td>1454000</td>\n",
       "      <td>0.828398</td>\n",
       "      <td>269.601925</td>\n",
       "      <td>256</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.651</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1458000</td>\n",
       "      <td>1459000</td>\n",
       "      <td>-0.086128</td>\n",
       "      <td>270.583243</td>\n",
       "      <td>272</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.651</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1463000</td>\n",
       "      <td>1464000</td>\n",
       "      <td>2.948188</td>\n",
       "      <td>236.321790</td>\n",
       "      <td>191</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.651</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556241</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137262000</td>\n",
       "      <td>137263000</td>\n",
       "      <td>-2.396149</td>\n",
       "      <td>263.131293</td>\n",
       "      <td>302</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.057053</td>\n",
       "      <td>0.063434</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.243724</td>\n",
       "      <td>-1.197675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556242</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137268000</td>\n",
       "      <td>137269000</td>\n",
       "      <td>3.640544</td>\n",
       "      <td>272.046527</td>\n",
       "      <td>212</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.057053</td>\n",
       "      <td>0.063434</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.243724</td>\n",
       "      <td>-1.197675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556243</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137269000</td>\n",
       "      <td>137270000</td>\n",
       "      <td>5.276351</td>\n",
       "      <td>300.458958</td>\n",
       "      <td>209</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.057053</td>\n",
       "      <td>0.063434</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.243724</td>\n",
       "      <td>-1.197675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556244</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137275000</td>\n",
       "      <td>137276000</td>\n",
       "      <td>2.687348</td>\n",
       "      <td>339.517045</td>\n",
       "      <td>290</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>1.564889</td>\n",
       "      <td>0.024254</td>\n",
       "      <td>0</td>\n",
       "      <td>0.194484</td>\n",
       "      <td>-1.615221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556245</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137290000</td>\n",
       "      <td>137291000</td>\n",
       "      <td>2.162811</td>\n",
       "      <td>257.721105</td>\n",
       "      <td>223</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>16.626887</td>\n",
       "      <td>0.384797</td>\n",
       "      <td>0</td>\n",
       "      <td>1.220811</td>\n",
       "      <td>-0.414769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1024014 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end   gnocchi  N_expected  N_observed  \\\n",
       "0        chr1    1432000    1433000  4.299894  338.059552         259   \n",
       "4        chr1    1451000    1452000  0.666316  302.590657         291   \n",
       "5        chr1    1453000    1454000  0.828398  269.601925         256   \n",
       "7        chr1    1458000    1459000 -0.086128  270.583243         272   \n",
       "11       chr1    1463000    1464000  2.948188  236.321790         191   \n",
       "...       ...        ...        ...       ...         ...         ...   \n",
       "1556241  chr9  137262000  137263000 -2.396149  263.131293         302   \n",
       "1556242  chr9  137268000  137269000  3.640544  272.046527         212   \n",
       "1556243  chr9  137269000  137270000  5.276351  300.458958         209   \n",
       "1556244  chr9  137275000  137276000  2.687348  339.517045         290   \n",
       "1556245  chr9  137290000  137291000  2.162811  257.721105         223   \n",
       "\n",
       "         window overlaps enhancer  window overlaps merged_exon      B  \\\n",
       "0                            True                        False  0.653   \n",
       "4                           False                        False  0.652   \n",
       "5                           False                        False  0.651   \n",
       "7                           False                        False  0.651   \n",
       "11                           True                        False  0.651   \n",
       "...                           ...                          ...    ...   \n",
       "1556241                      True                        False  0.536   \n",
       "1556242                      True                        False  0.536   \n",
       "1556243                      True                        False  0.536   \n",
       "1556244                      True                        False  0.536   \n",
       "1556245                     False                        False  0.536   \n",
       "\n",
       "         paternal_recombination_rate  maternal_recombination_rate  \\\n",
       "0                           0.039046                     0.025852   \n",
       "4                           0.039046                     0.025852   \n",
       "5                           0.039046                     0.025852   \n",
       "7                           0.039046                     0.025852   \n",
       "11                          0.039046                     0.025852   \n",
       "...                              ...                          ...   \n",
       "1556241                     0.057053                     0.063434   \n",
       "1556242                     0.057053                     0.063434   \n",
       "1556243                     0.057053                     0.063434   \n",
       "1556244                     1.564889                     0.024254   \n",
       "1556245                    16.626887                     0.384797   \n",
       "\n",
       "         gBGC-tract_overlap_count  log10_paternal_recombination_rate  \\\n",
       "0                               0                          -1.408422   \n",
       "4                               0                          -1.408422   \n",
       "5                               0                          -1.408422   \n",
       "7                               0                          -1.408422   \n",
       "11                              0                          -1.408422   \n",
       "...                           ...                                ...   \n",
       "1556241                         0                          -1.243724   \n",
       "1556242                         0                          -1.243724   \n",
       "1556243                         0                          -1.243724   \n",
       "1556244                         0                           0.194484   \n",
       "1556245                         0                           1.220811   \n",
       "\n",
       "         log10_maternal_recombination_rate  \n",
       "0                                -1.587499  \n",
       "4                                -1.587499  \n",
       "5                                -1.587499  \n",
       "7                                -1.587499  \n",
       "11                               -1.587499  \n",
       "...                                    ...  \n",
       "1556241                          -1.197675  \n",
       "1556242                          -1.197675  \n",
       "1556243                          -1.197675  \n",
       "1556244                          -1.615221  \n",
       "1556245                          -0.414769  \n",
       "\n",
       "[1024014 rows x 14 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_header(filename):\n",
    "  with open(filename) as fh: \n",
    "    header = fh.readline().strip().split('\\t')\n",
    "  return header\n",
    "\n",
    "def intersect(a_filename, b_filename, intersect_filename, b_feature): \n",
    "  cmd = (\n",
    "    f'bedtools intersect'\n",
    "    f\" -a <(tail -n +2 {a_filename})\" # contains header\n",
    "    f\" -b <(cat {b_filename})\"        # does not contain header\n",
    "    f' -wao -f 0.5'\n",
    "    f' > {intersect_filename}'\n",
    "  )\n",
    "  shell(cmd)\n",
    "  print_string_as_info(f'Wrote {intersect_filename}')\n",
    "\n",
    "  a_file_header = get_header(a_filename) # contains header\n",
    "  b_file_header = [f'{b_feature}_chrom', f'{b_feature}_start', f'{b_feature}_end', b_feature] # does not contain header\n",
    "\n",
    "  return a_file_header, b_file_header\n",
    "\n",
    "def count_overlaps(a_filename, b_filename, count_filename, b_feature): \n",
    "  cmd = (\n",
    "    f'bedtools intersect'\n",
    "    f\" -a <(tail -n +2 {a_filename})\" # contains header\n",
    "    f\" -b <(cat {b_filename})\"        # does not contain header\n",
    "    f' -c'                            # count overlaps\n",
    "    f' > {count_filename}'\n",
    "  )\n",
    "  shell(cmd)\n",
    "  print_string_as_info(f'Wrote {count_filename}')\n",
    "\n",
    "  return get_header(a_filename) + [f'{b_feature}_overlap_count']\n",
    "\n",
    "def set_column_dtypes(df, column_dtypes, log=False): \n",
    "  for col, dtype in column_dtypes.items(): \n",
    "    df[col] = df[col].astype(dtype)\n",
    "\n",
    "  if log: print(dict(df.dtypes))\n",
    "  return df \n",
    "\n",
    "def make_scores_numeric(df, b_feature): \n",
    "  # we don't want to include enhancers that don't overlap any windows with b_feature values\n",
    "  df = df[df[f'{b_feature}_chrom'] != '.'] \n",
    "\n",
    "  # having removed records where 'b_feature_chrom' is '.', we can now convert the 'b_feature' column to a float\n",
    "  df = set_column_dtypes(\n",
    "    df,   \n",
    "    column_dtypes = {\n",
    "      b_feature: float,\n",
    "    }\n",
    "  )\n",
    "\n",
    "  return df \n",
    "  \n",
    "def intersect_and_aggregate(a_filename_stem, b_filename, b_feature, b_feature_aggregation): \n",
    "  a_filename = f'{a_filename_stem}.bed'\n",
    "  intersect_filename = f'{a_filename_stem}.intersect.{b_feature}-map.bed'\n",
    "\n",
    "  a_file_header, b_file_header = intersect(a_filename, b_filename, intersect_filename, b_feature)\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    intersect_filename, \n",
    "    sep='\\t',\n",
    "    names=a_file_header + b_file_header + [f'chenWindow_{b_feature}Window_overlap'],\n",
    "  )\n",
    "\n",
    "  df = make_scores_numeric(df, b_feature)\n",
    "\n",
    "  # some Chen-windows may intersect multiple b_feature-windows, \n",
    "  # so let's group by Chen-window, and aggregate scores over all b_feature-windows in the group\n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = a_file_header,\n",
    "    aggregation_functions = {\n",
    "      b_feature: [b_feature_aggregation],\n",
    "    }\n",
    "  )\n",
    "  df = df.rename(columns={f'{b_feature_aggregation} {b_feature}': b_feature})\n",
    "\n",
    "  df.to_csv(\n",
    "    f'{a_filename_stem}.{b_feature}.bed',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "  )\n",
    "\n",
    "def count_overlaps_wrapper(a_filename_stem, b_filename, b_feature): \n",
    "  a_filename = f'{a_filename_stem}.bed'\n",
    "  count_filename = f'{a_filename_stem}.{b_feature}-counts.bed'\n",
    "\n",
    "  count_file_header = count_overlaps(a_filename, b_filename, count_filename, b_feature)\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    count_filename, \n",
    "    sep='\\t',\n",
    "    names=count_file_header,\n",
    "  )\n",
    "\n",
    "  df.to_csv(\n",
    "    count_filename,\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "  )\n",
    "\n",
    "def assign_scores_to_chen_windows():\n",
    "  df_gnocchi = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2-gnocchi.bed',\n",
    "    # f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2-gnocchi.test.bed', # TODO: comment out this line\n",
    "    sep='\\t',\n",
    "    names=['chrom', 'start', 'end', 'gnocchi']\n",
    "  )\n",
    "  df_N_expected = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2-N_expected_gnocchi.bed',\n",
    "    sep='\\t',\n",
    "    names=['chrom', 'start', 'end', 'N_expected']\n",
    "  )\n",
    "  df_N_observed = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2-N_observed.bed',\n",
    "    sep='\\t',\n",
    "    names=['chrom', 'start', 'end', 'N_observed']\n",
    "  )\n",
    "  df_enhancers_exons = WINDOWS_0\n",
    "\n",
    "  df = pd.merge(\n",
    "    df_gnocchi, \n",
    "    df_N_expected, \n",
    "    on=['chrom', 'start', 'end'],\n",
    "    how='inner'\n",
    "  )\n",
    "  df = pd.merge(\n",
    "    df, \n",
    "    df_N_observed, \n",
    "    on=['chrom', 'start', 'end'],\n",
    "    how='inner'\n",
    "  )\n",
    "  df = pd.merge(\n",
    "    df, \n",
    "    df_enhancers_exons, \n",
    "    on=['chrom', 'start', 'end'],\n",
    "    how='inner'\n",
    "  )\n",
    "\n",
    "  df.to_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.bed',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "  )\n",
    "\n",
    "  # assign BGS values to Chen et al. windows \n",
    "  intersect_and_aggregate(\n",
    "    a_filename_stem = f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed', \n",
    "    b_filename = f'{CONSTRAINT_TOOLS_DATA}/background-selection/CADD-B-map/bmap.hg38.bed', \n",
    "    b_feature = 'B', \n",
    "    b_feature_aggregation = 'max'\n",
    "  )   \n",
    "\n",
    "  # assign paternal recombination rates to Chen et al. windows \n",
    "  intersect_and_aggregate(\n",
    "    a_filename_stem = f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B', \n",
    "    b_filename = f'{CONSTRAINT_TOOLS_DATA}/GC-biased-gene-conversion/paternal-recombination-rate.grch38.bed', \n",
    "    b_feature = 'paternal_recombination_rate', \n",
    "    b_feature_aggregation = 'mean',\n",
    "  )   \n",
    "\n",
    "  # assign maternal recombination rates to Chen et al. windows \n",
    "  intersect_and_aggregate(\n",
    "    a_filename_stem = f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate', \n",
    "    b_filename = f'{CONSTRAINT_TOOLS_DATA}/GC-biased-gene-conversion/maternal-recombination-rate.grch38.bed', \n",
    "    b_feature = 'maternal_recombination_rate', \n",
    "    b_feature_aggregation = 'mean'\n",
    "  )   \n",
    "\n",
    "  # compute number of gBGC tracts that each Chen et al. window overlaps\n",
    "  count_overlaps_wrapper(\n",
    "    a_filename_stem = f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.maternal_recombination_rate', \n",
    "    b_filename = f'{CONSTRAINT_TOOLS_DATA}/GC-biased-gene-conversion/gBGC-tracts.hg38.bed', \n",
    "    b_feature = 'gBGC-tract'\n",
    "  )\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.maternal_recombination_rate.gBGC-tract-counts.bed',\n",
    "    sep='\\t',\n",
    "  )\n",
    "\n",
    "  df = df[df['window overlaps merged_exon'] == False] # strictly non-exonic windows\n",
    "\n",
    "  # throw out windows with recombination rates less than 1e-8, and take the log10 of the recombination rates\n",
    "  for sex in ['paternal', 'maternal']:\n",
    "    df = df[df[f'{sex}_recombination_rate'] > 1e-8] # log10(0) is undefined \n",
    "    df[f'log10_{sex}_recombination_rate'] = np.log10(df[f'{sex}_recombination_rate'])\n",
    "\n",
    "  df.to_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.maternal_recombination_rate.gBGC-tract-counts.non-exonic.bed',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "  )\n",
    "\n",
    "  return df\n",
    "\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "WINDOWS = assign_scores_to_chen_windows()\n",
    "WINDOWS"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "constraint-tools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
