{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "CONSTRAINT_TOOLS = '/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools'\n",
    "CONSTRAINT_TOOLS_DATA = '/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools'\n",
    "\n",
    "import sys\n",
    "sys.path.append(f'{CONSTRAINT_TOOLS}/utilities')\n",
    "\n",
    "LOWER_DELETION_SIZE_LIMIT = 4000\n",
    "UPPER_DELETION_SIZE_LIMIT = 10000\n",
    "DELETION_ALLELE_FREQ_THRESHOLD = 0.00001\n",
    "\n",
    "# DELETION_TYPE = 'long' \n",
    "DELETION_TYPE = 'all' \n",
    "\n",
    "BALANCE_CLASSES = False\n",
    "\n",
    "PUBLIC_REPO = 'constraint-tools-data'\n",
    "PUBLIC_REPO_DIR = f'/scratch/ucgd/lustre-work/quinlan/u6018199/{PUBLIC_REPO}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute SNV scores of a set of positive and negative genes (i.e. genes that are known to be under negative selection or no selection) by using ONLY the exons of these genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/benchmark-genome-wide-predictions/chen-et-al-2022/negative-canonical-exons-intersect-chen-windows.bed\u001b[0m\n",
      "<ipython-input-2-de7c68da8dc5>:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[col] = df[col].astype(dtype)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>min negative new chen zscore over all exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR10A3</td>\n",
       "      <td>-0.817755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR10A4</td>\n",
       "      <td>1.320257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR10A5</td>\n",
       "      <td>-0.349816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR10A6</td>\n",
       "      <td>2.569226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR10AD1</td>\n",
       "      <td>0.050568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>OR9G4</td>\n",
       "      <td>3.281551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>OR9I1</td>\n",
       "      <td>0.853197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>OR9K2</td>\n",
       "      <td>3.569851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>OR9Q1</td>\n",
       "      <td>-0.223640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>OR9Q2</td>\n",
       "      <td>1.714794</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>268 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_symbol  min negative new chen zscore over all exons\n",
       "0        OR10A3                                    -0.817755\n",
       "1        OR10A4                                     1.320257\n",
       "2        OR10A5                                    -0.349816\n",
       "3        OR10A6                                     2.569226\n",
       "4       OR10AD1                                     0.050568\n",
       "..          ...                                          ...\n",
       "263       OR9G4                                     3.281551\n",
       "264       OR9I1                                     0.853197\n",
       "265       OR9K2                                     3.569851\n",
       "266       OR9Q1                                    -0.223640\n",
       "267       OR9Q2                                     1.714794\n",
       "\n",
       "[268 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from shell import shell \n",
    "from colorize import print_string_as_info\n",
    "from aggregate import aggregate\n",
    "\n",
    "def get_header(filename):\n",
    "  with open(filename) as fh: \n",
    "    header = fh.readline().strip().split('\\t')\n",
    "  return header\n",
    "\n",
    "def intersect(a_filename, b_filename, intersect_filename): \n",
    "  cmd = (\n",
    "    f'bedtools intersect'\n",
    "    f\" -a <(tail -n +2 {a_filename})\" \n",
    "    f\" -b <(tail -n +2 {b_filename})\"\n",
    "    f' -wao -f 0.5'\n",
    "    f' > {intersect_filename}'\n",
    "  )\n",
    "  shell(cmd)\n",
    "  print_string_as_info(f'Wrote {intersect_filename}')\n",
    "\n",
    "  a_file_header = get_header(a_filename)\n",
    "  b_file_header = get_header(b_filename)\n",
    "\n",
    "  return a_file_header, b_file_header\n",
    "\n",
    "def set_column_dtypes(df, column_dtypes, log=False): \n",
    "  for col, dtype in column_dtypes.items(): \n",
    "    df[col] = df[col].astype(dtype)\n",
    "\n",
    "  if log: print(dict(df.dtypes))\n",
    "  return df \n",
    "\n",
    "def make_SNV_score_numeric(df): \n",
    "  # we don't want to include exons that don't overlap any Chen windows\n",
    "  df = df[df['chrom'] != '.'] \n",
    "\n",
    "  # having removed records where 'negative new chen zscore' is '.', we can now convert the column to a float\n",
    "  df = set_column_dtypes(\n",
    "    df,   \n",
    "    column_dtypes = {\n",
    "      'negative new chen zscore': float,\n",
    "    }\n",
    "  )\n",
    "\n",
    "  return df \n",
    "  \n",
    "def assign_SNV_scores_to_labeled_genes(gene_class):\n",
    "  a_filename = f'{CONSTRAINT_TOOLS_DATA}/genes/grch38/canonical-exons.{gene_class}.sorted.bed'\n",
    "  b_filename = f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/small-windows.bed'\n",
    "  intersect_filename = f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/{gene_class}-canonical-exons-intersect-chen-windows.bed'\n",
    "\n",
    "  a_file_header, b_file_header = intersect(a_filename, b_filename, intersect_filename)\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    intersect_filename, \n",
    "    sep='\\t',\n",
    "    names=a_file_header + b_file_header + ['exon_chen_overlap'],\n",
    "  )\n",
    "\n",
    "  df = df[\n",
    "    (df['chromosome'] != 'chrX') &\n",
    "    (df['chromosome'] != 'chrY')\n",
    "  ]\n",
    "\n",
    "  df = make_SNV_score_numeric(df)\n",
    "  \n",
    "  # though unlikely, some exons may intersect multiple Chen windows, \n",
    "  # so let's group by exon, and compute the min Chen score over all Chen windows in the group\n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = a_file_header,\n",
    "    aggregation_functions = {\n",
    "      'negative new chen zscore': [np.min],\n",
    "    }\n",
    "  )\n",
    "\n",
    "  # group exons by parent gene, computing min z-score in each group\n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = ['gene_symbol'],\n",
    "    aggregation_functions = {\n",
    "      'min negative new chen zscore': [np.min],\n",
    "    }\n",
    "  )\n",
    "\n",
    "  # [TODO] possibly normalize by number of exons in each gene (this would make sense if the aggregate were a mean, but it's a min, so maybe not)\n",
    "\n",
    "  df = df.rename(columns={\n",
    "    'min min negative new chen zscore': 'min negative new chen zscore over all exons'\n",
    "  })\n",
    "\n",
    "  return df \n",
    "\n",
    "assign_SNV_scores_to_labeled_genes(gene_class='negative')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute SV scores of a set of positive and negative genes (i.e. genes that are known to be under negative selection or no selection) by using ONLY the exons of these genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;36mWe assume that the first line of the following is a header line: \u001b[0m/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/genes/grch38/canonical-exons.positive.sorted.bed\n",
      "\u001b[0;36mWrote windows with deletion counts and overlaps to: \u001b[0m/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/benchmark-genome-wide-predictions/chen-et-al-2022/positive-canonical-exons-intersect-all-observed-deletions.bed\n",
      "\u001b[0;36mWrote all observed deletions in UCSC-genome-browser format to: \u001b[0m/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools-data/all-observed-deletions.bed\n",
      "Everything up-to-date\n",
      "\u001b[0;36mPushed /scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools-data/all-observed-deletions.bed to public repo\u001b[0m\n",
      "On branch main\n",
      "Your branch is up to date with 'origin/main'.\n",
      "\n",
      "nothing to commit, working tree clean\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>exon_count</th>\n",
       "      <th>deletion count over all exons</th>\n",
       "      <th>deletion overlap over all exons</th>\n",
       "      <th>total length of all exons</th>\n",
       "      <th>gene coordinates</th>\n",
       "      <th>deletion count per bp</th>\n",
       "      <th>fraction of bps that are deleted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTN2</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>407</td>\n",
       "      <td>4851</td>\n",
       "      <td>chr1:236686499-236764631</td>\n",
       "      <td>0.000618</td>\n",
       "      <td>0.083900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADAR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6595</td>\n",
       "      <td>chr1:154582057-154608186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AHDC1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6431</td>\n",
       "      <td>chr1:27534303-27603632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMPD1</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>866</td>\n",
       "      <td>2359</td>\n",
       "      <td>chr1:114673098-114695586</td>\n",
       "      <td>0.002543</td>\n",
       "      <td>0.367105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APOA2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>470</td>\n",
       "      <td>chr1:161222292-161223628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>TGFBR1</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>869</td>\n",
       "      <td>5924</td>\n",
       "      <td>chr9:99105150-99153658</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.146691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>TOR1A</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2075</td>\n",
       "      <td>chr9:129812942-129824136</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>TPM2</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>1318</td>\n",
       "      <td>1318</td>\n",
       "      <td>chr9:35682926-35690056</td>\n",
       "      <td>0.007587</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "      <td>3792</td>\n",
       "      <td>8593</td>\n",
       "      <td>chr9:132891348-132944633</td>\n",
       "      <td>0.000698</td>\n",
       "      <td>0.441289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>VCP</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3729</td>\n",
       "      <td>chr9:35056064-35072625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>646 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_symbol  exon_count  deletion count over all exons  \\\n",
       "0         ACTN2          21                              3   \n",
       "1          ADAR          15                              0   \n",
       "2         AHDC1           7                              0   \n",
       "3         AMPD1          16                              6   \n",
       "4         APOA2           4                              0   \n",
       "..          ...         ...                            ...   \n",
       "641      TGFBR1           9                              4   \n",
       "642       TOR1A           5                              0   \n",
       "643        TPM2           9                             10   \n",
       "644        TSC1          23                              6   \n",
       "645         VCP          17                              0   \n",
       "\n",
       "     deletion overlap over all exons  total length of all exons  \\\n",
       "0                                407                       4851   \n",
       "1                                  0                       6595   \n",
       "2                                  0                       6431   \n",
       "3                                866                       2359   \n",
       "4                                  0                        470   \n",
       "..                               ...                        ...   \n",
       "641                              869                       5924   \n",
       "642                                0                       2075   \n",
       "643                             1318                       1318   \n",
       "644                             3792                       8593   \n",
       "645                                0                       3729   \n",
       "\n",
       "             gene coordinates  deletion count per bp  \\\n",
       "0    chr1:236686499-236764631               0.000618   \n",
       "1    chr1:154582057-154608186               0.000000   \n",
       "2      chr1:27534303-27603632               0.000000   \n",
       "3    chr1:114673098-114695586               0.002543   \n",
       "4    chr1:161222292-161223628               0.000000   \n",
       "..                        ...                    ...   \n",
       "641    chr9:99105150-99153658               0.000675   \n",
       "642  chr9:129812942-129824136               0.000000   \n",
       "643    chr9:35682926-35690056               0.007587   \n",
       "644  chr9:132891348-132944633               0.000698   \n",
       "645    chr9:35056064-35072625               0.000000   \n",
       "\n",
       "     fraction of bps that are deleted  \n",
       "0                            0.083900  \n",
       "1                            0.000000  \n",
       "2                            0.000000  \n",
       "3                            0.367105  \n",
       "4                            0.000000  \n",
       "..                                ...  \n",
       "641                          0.146691  \n",
       "642                          0.000000  \n",
       "643                          1.000000  \n",
       "644                          0.441289  \n",
       "645                          0.000000  \n",
       "\n",
       "[646 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def assign_SV_scores_to_labeled_genes(\n",
    "    gene_class, \n",
    "    deletion_class,\n",
    "    deletion_type=DELETION_TYPE,\n",
    "):\n",
    "  windows = f'{CONSTRAINT_TOOLS_DATA}/genes/grch38/canonical-exons.{gene_class}.sorted.bed'\n",
    "  windows_with_deletions = f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/{gene_class}-canonical-exons-intersect-{deletion_type}-{deletion_class}-deletions.bed'\n",
    "\n",
    "  intersect_script = f'{CONSTRAINT_TOOLS}/experiments/germline-model/chen-et-al-2022/intersect-windows-with-deletions.sh'\n",
    "\n",
    "  PROCESSED_DELETIONS = f'{deletion_type}-{deletion_class}-deletions.bed'\n",
    "\n",
    "  cmd = (\n",
    "    f'bash '\n",
    "    f'{intersect_script} '\n",
    "    f'{windows} {deletion_class} {windows_with_deletions} '\n",
    "    f'{PUBLIC_REPO_DIR} {PROCESSED_DELETIONS} '\n",
    "    f'{deletion_type} '\n",
    "    f'{LOWER_DELETION_SIZE_LIMIT} {UPPER_DELETION_SIZE_LIMIT} {DELETION_ALLELE_FREQ_THRESHOLD}'\n",
    "  )\n",
    "  print(shell(cmd))\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    windows_with_deletions, \n",
    "    sep='\\t',\n",
    "  )\n",
    "\n",
    "  df = df[\n",
    "    (df['chromosome'] != 'chrX') &\n",
    "    (df['chromosome'] != 'chrY')\n",
    "  ]\n",
    "\n",
    "  # though unlikely, some exons may intersect multiple merged (and therefore non-overlapping) deletions,\n",
    "  # therefore group by exon (and deletion count), and compute \"sum deletion overlap\" in each group,\n",
    "  group_columns = df.columns.tolist()\n",
    "  group_columns.remove('deletion_overlap')\n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = group_columns,\n",
    "    aggregation_functions = {\n",
    "      'deletion_overlap': [np.sum],\n",
    "    }\n",
    "  )\n",
    "\n",
    "  df['exon_length'] = df['exon_end'] - df['exon_start']\n",
    "\n",
    "  # group exons by parent gene, \n",
    "  # computing \"sum deletion count\" and \"sum deletion overlap\" in each group, \n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = ['chromosome', 'gene_symbol'],\n",
    "    aggregation_functions = {\n",
    "      'exon_start': [np.min],\n",
    "      'exon_end': [np.max],\n",
    "      'exon_rank': 'count',\n",
    "      'deletion_count': [np.sum],\n",
    "      'sum deletion_overlap': [np.sum],\n",
    "      'exon_length': [np.sum],\n",
    "    }\n",
    "  )\n",
    "\n",
    "  df['gene coordinates'] = df['chromosome'] + ':' + df['min exon_start'].astype(str) + '-' + df['max exon_end'].astype(str)\n",
    "  df = df.drop(columns=['chromosome', 'min exon_start', 'max exon_end'])\n",
    "  df['deletion count per bp'] = df['sum deletion_count'] / df['sum exon_length']\n",
    "  df['fraction of bps that are deleted'] = df['sum sum deletion_overlap'] / df['sum exon_length']\n",
    "\n",
    "  df = df.rename(columns={\n",
    "    'count exon_rank': 'exon_count',\n",
    "    'sum deletion_count': 'deletion count over all exons',\n",
    "    'sum sum deletion_overlap': 'deletion overlap over all exons',\n",
    "    'sum exon_length': 'total length of all exons',\n",
    "  })\n",
    "\n",
    "  return df\n",
    "\n",
    "assign_SV_scores_to_labeled_genes(\n",
    "  gene_class='positive',\n",
    "  deletion_class='observed'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: the null statistics are computed using shuffled versions of the topmed deletions\n",
    "# 2.2 copy experiments/germline-model/chen-et-al-2022/get-observed-deletions.sh \n",
    "#     write a bash function called get-deletions-tail that shuffles the topmed deletions: bedtools shuffle -i <(get-topmed-deletions-tail) -g my.genome -excl <(get-exclude-regions) \n",
    "#     write a bash function called get-deletion-head that returns the TOPMED header \n",
    "# 2.3 validate by running assign_SV_scores_to_labeled_genes with deletion_class='shuffled'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 3. wrap the python code that calls intersect-windows-with-deletions.sh in a function \n",
    "#    that takes as argument get-deletions, and PROCESSED_DELETIONS,\n",
    "#    returns the fraction of (negative or positive) genes that have no or completely overlapping deletions, \n",
    "#    and then call the function 1 time with get-deletions==get-observed-deletions to get the observed fraction,\n",
    "#    and 1000 times with get-deletions==get-shuffled-deletions to get the null distribution of fractions\n",
    "# 4. plot the distribution of fractions under the null, together with the observed fraction\n",
    "# 5. compute the p-value as the fraction of null fractions that are greater than the observed fraction\n",
    "# 6. do all the above for gene-class==negative (expecting the pvalue to be insignificant) and gene-class==positive (expecting the pvalue to be significant)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine SNV and SV scores of positive and negative genes \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_examples(gene_class): \n",
    "  df = pd.merge(\n",
    "    assign_SNV_scores_to_labeled_genes(gene_class),\n",
    "    assign_SV_scores_to_labeled_genes(gene_class),\n",
    "    on='gene_symbol',\n",
    "    how='inner',\n",
    "  )\n",
    "  df['truly constrained'] = gene_class == 'positive'\n",
    "  return df\n",
    "\n",
    "def concat(dfs): \n",
    "  df = pd.concat(dfs)\n",
    "  df = df.reset_index(drop=True) # drop duplicates in index so that later index-based operations are possible \n",
    "  return df\n",
    "\n",
    "def label_examples(): \n",
    "  negative_examples = create_examples(gene_class='negative')\n",
    "  positive_examples = create_examples(gene_class='positive')\n",
    "\n",
    "  if BALANCE_CLASSES: # balance classes for later training \n",
    "    positive_examples = positive_examples.sample(n=len(negative_examples), random_state=0) \n",
    "\n",
    "  df = concat([negative_examples, positive_examples])\n",
    "\n",
    "  return df \n",
    "\n",
    "LABELED_EXAMPLES = label_examples()\n",
    "LABELED_EXAMPLES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the distribution of SNV and SV scores of positive and negative genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "\n",
    "# Set the font size\n",
    "mpl.rcParams['font.size'] = 30\n",
    "plt.rcParams['figure.figsize'] = (10, 8)\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "plt.figure(figsize=(10, 8))\n",
    "sns.histplot(data=LABELED_EXAMPLES, x='min negative new chen zscore over all exons', hue='truly constrained', kde=True)\n",
    "plt.xlabel('SNV score')\n",
    "plt.ylabel('Number of genes')\n",
    "plt.show()\n",
    "\n",
    "def plot_distribution(ser, label): \n",
    "  counts, edges = np.histogram(ser, bins=20)\n",
    "  total_count = np.sum(counts)\n",
    "  centers = (edges[:-1] + edges[1:]) / 2\n",
    "  # plt.plot(centers, counts/total_count, label=label, alpha=1)\n",
    "  plt.plot(centers, counts, label=label, alpha=1)\n",
    "\n",
    "def filter_scores_on_constraint(df, constrained, score): \n",
    "  df = df[df['truly constrained'] == constrained]\n",
    "  ser = df[score]\n",
    "  return ser\n",
    "\n",
    "def plot_distributions(df, score, score_alias):\n",
    "  plot_distribution(filter_scores_on_constraint(df, constrained=False, score=score), label='negative genes')\n",
    "  plot_distribution(filter_scores_on_constraint(df, constrained=True, score=score), label='positive genes')\n",
    "  plt.legend()\n",
    "  plt.xlabel(score_alias)\n",
    "  # plt.ylabel('Fraction of genes')\n",
    "  plt.ylabel('Number of genes')\n",
    "  plt.show()\n",
    "\n",
    "plot_distributions(\n",
    "  LABELED_EXAMPLES, \n",
    "  score='deletion count per bp',\n",
    "  score_alias='SV score (deletion count per bp)'\n",
    ")\n",
    "\n",
    "plot_distributions(\n",
    "  LABELED_EXAMPLES, \n",
    "  score='fraction of bps that are deleted',\n",
    "  score_alias='SV score (fraction of bps that are deleted)'\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "constraint-tools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
