{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1d625122",
   "metadata": {},
   "source": [
    "## Fetch windows, zscores, annotations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "81784063",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width: 100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "display(HTML(\"<style>.container { width: 100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf7bb755",
   "metadata": {},
   "outputs": [],
   "source": [
    "CONSTRAINT_TOOLS = '/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools'\n",
    "CONSTRAINT_TOOLS_DATA = '/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bfcae58a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>N_observed</th>\n",
       "      <th>N_bar_7_noncoding</th>\n",
       "      <th>N_mean_null_7_noncoding</th>\n",
       "      <th>N_variance_null_7_noncoding</th>\n",
       "      <th>N_bar_1_noncoding</th>\n",
       "      <th>N_mean_null_1_noncoding</th>\n",
       "      <th>N_variance_null_1_noncoding</th>\n",
       "      <th>N_bar_1_noncoding_noisy</th>\n",
       "      <th>N_mean_null_1_noncoding_noisy</th>\n",
       "      <th>N_variance_null_1_noncoding_noisy</th>\n",
       "      <th>N_bar_3_noncoding</th>\n",
       "      <th>N_mean_null_3_noncoding</th>\n",
       "      <th>N_variance_null_3_noncoding</th>\n",
       "      <th>N_bar_3_noncoding_noisy</th>\n",
       "      <th>N_mean_null_3_noncoding_noisy</th>\n",
       "      <th>N_variance_null_3_noncoding_noisy</th>\n",
       "      <th>N_bar_5_noncoding</th>\n",
       "      <th>N_mean_null_5_noncoding</th>\n",
       "      <th>N_variance_null_5_noncoding</th>\n",
       "      <th>N_bar_5_noncoding_noisy</th>\n",
       "      <th>N_mean_null_5_noncoding_noisy</th>\n",
       "      <th>N_variance_null_5_noncoding_noisy</th>\n",
       "      <th>enhancer overlap</th>\n",
       "      <th>merged_exon overlap</th>\n",
       "      <th>window overlaps enhancer</th>\n",
       "      <th>window overlaps merged_exon</th>\n",
       "      <th>window overlaps (enhancer, merged_exon)</th>\n",
       "      <th>new chen zscore</th>\n",
       "      <th>positive-vista-enhancer count</th>\n",
       "      <th>delta_sf</th>\n",
       "      <th>delta_f</th>\n",
       "      <th>window overlaps positive vista enhancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1432000</td>\n",
       "      <td>1433000</td>\n",
       "      <td>260</td>\n",
       "      <td>-0.372778</td>\n",
       "      <td>265.321432</td>\n",
       "      <td>203.777753</td>\n",
       "      <td>1.590921</td>\n",
       "      <td>236.519249</td>\n",
       "      <td>217.834403</td>\n",
       "      <td>1.676924</td>\n",
       "      <td>235.282294</td>\n",
       "      <td>217.264932</td>\n",
       "      <td>-0.247615</td>\n",
       "      <td>263.528599</td>\n",
       "      <td>203.072610</td>\n",
       "      <td>-0.283079</td>\n",
       "      <td>264.038617</td>\n",
       "      <td>203.540686</td>\n",
       "      <td>-0.339806</td>\n",
       "      <td>264.852332</td>\n",
       "      <td>203.910737</td>\n",
       "      <td>-0.340167</td>\n",
       "      <td>264.858613</td>\n",
       "      <td>204.004160</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>(True, False)</td>\n",
       "      <td>4.299894</td>\n",
       "      <td>0</td>\n",
       "      <td>-5.321432</td>\n",
       "      <td>1.792832</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1435000</td>\n",
       "      <td>1436000</td>\n",
       "      <td>275</td>\n",
       "      <td>-6.280673</td>\n",
       "      <td>368.627875</td>\n",
       "      <td>222.227572</td>\n",
       "      <td>2.097960</td>\n",
       "      <td>243.596443</td>\n",
       "      <td>224.059444</td>\n",
       "      <td>2.193623</td>\n",
       "      <td>242.210058</td>\n",
       "      <td>223.438246</td>\n",
       "      <td>-7.547130</td>\n",
       "      <td>387.939429</td>\n",
       "      <td>223.937877</td>\n",
       "      <td>-7.609011</td>\n",
       "      <td>389.032484</td>\n",
       "      <td>224.595337</td>\n",
       "      <td>-6.810107</td>\n",
       "      <td>375.796139</td>\n",
       "      <td>219.068492</td>\n",
       "      <td>-6.784165</td>\n",
       "      <td>375.458319</td>\n",
       "      <td>219.269877</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>444.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(True, True)</td>\n",
       "      <td>3.331645</td>\n",
       "      <td>0</td>\n",
       "      <td>-93.627875</td>\n",
       "      <td>-19.311554</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1449000</td>\n",
       "      <td>1450000</td>\n",
       "      <td>225</td>\n",
       "      <td>0.139977</td>\n",
       "      <td>223.137143</td>\n",
       "      <td>177.111728</td>\n",
       "      <td>0.959620</td>\n",
       "      <td>211.569144</td>\n",
       "      <td>195.888493</td>\n",
       "      <td>1.011359</td>\n",
       "      <td>210.858989</td>\n",
       "      <td>195.501382</td>\n",
       "      <td>0.258359</td>\n",
       "      <td>221.559138</td>\n",
       "      <td>177.372370</td>\n",
       "      <td>0.225283</td>\n",
       "      <td>221.996472</td>\n",
       "      <td>177.748203</td>\n",
       "      <td>0.330975</td>\n",
       "      <td>220.612316</td>\n",
       "      <td>175.743926</td>\n",
       "      <td>0.330023</td>\n",
       "      <td>220.624461</td>\n",
       "      <td>175.781765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>311.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>(False, True)</td>\n",
       "      <td>1.817267</td>\n",
       "      <td>0</td>\n",
       "      <td>1.862857</td>\n",
       "      <td>1.578005</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1450000</td>\n",
       "      <td>1451000</td>\n",
       "      <td>313</td>\n",
       "      <td>2.524948</td>\n",
       "      <td>277.859484</td>\n",
       "      <td>193.691881</td>\n",
       "      <td>5.870332</td>\n",
       "      <td>227.882674</td>\n",
       "      <td>210.237742</td>\n",
       "      <td>5.950235</td>\n",
       "      <td>226.828073</td>\n",
       "      <td>209.731395</td>\n",
       "      <td>2.197534</td>\n",
       "      <td>281.968822</td>\n",
       "      <td>199.400099</td>\n",
       "      <td>2.147648</td>\n",
       "      <td>282.634862</td>\n",
       "      <td>199.905291</td>\n",
       "      <td>2.480911</td>\n",
       "      <td>278.322530</td>\n",
       "      <td>195.376527</td>\n",
       "      <td>2.482902</td>\n",
       "      <td>278.288535</td>\n",
       "      <td>195.445943</td>\n",
       "      <td>NaN</td>\n",
       "      <td>758.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>(False, True)</td>\n",
       "      <td>1.365863</td>\n",
       "      <td>0</td>\n",
       "      <td>35.140516</td>\n",
       "      <td>-4.109338</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1451000</td>\n",
       "      <td>1452000</td>\n",
       "      <td>293</td>\n",
       "      <td>2.951595</td>\n",
       "      <td>252.829443</td>\n",
       "      <td>185.226049</td>\n",
       "      <td>4.803930</td>\n",
       "      <td>223.924244</td>\n",
       "      <td>206.755939</td>\n",
       "      <td>4.877094</td>\n",
       "      <td>222.953222</td>\n",
       "      <td>206.278524</td>\n",
       "      <td>2.367303</td>\n",
       "      <td>260.131776</td>\n",
       "      <td>192.772432</td>\n",
       "      <td>2.332657</td>\n",
       "      <td>260.580640</td>\n",
       "      <td>193.155585</td>\n",
       "      <td>2.725116</td>\n",
       "      <td>255.650992</td>\n",
       "      <td>187.839972</td>\n",
       "      <td>2.724052</td>\n",
       "      <td>255.656435</td>\n",
       "      <td>187.931976</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>(False, False)</td>\n",
       "      <td>0.666316</td>\n",
       "      <td>0</td>\n",
       "      <td>40.170557</td>\n",
       "      <td>-7.302333</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786072</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137269000</td>\n",
       "      <td>137270000</td>\n",
       "      <td>210</td>\n",
       "      <td>-3.060479</td>\n",
       "      <td>251.650751</td>\n",
       "      <td>185.210985</td>\n",
       "      <td>-0.657779</td>\n",
       "      <td>219.366052</td>\n",
       "      <td>202.746590</td>\n",
       "      <td>-0.596997</td>\n",
       "      <td>218.491272</td>\n",
       "      <td>202.302491</td>\n",
       "      <td>-3.445370</td>\n",
       "      <td>257.657992</td>\n",
       "      <td>191.337298</td>\n",
       "      <td>-3.474260</td>\n",
       "      <td>258.105694</td>\n",
       "      <td>191.720447</td>\n",
       "      <td>-3.228932</td>\n",
       "      <td>254.283943</td>\n",
       "      <td>188.093965</td>\n",
       "      <td>-3.225423</td>\n",
       "      <td>254.247196</td>\n",
       "      <td>188.190706</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>(True, False)</td>\n",
       "      <td>5.276351</td>\n",
       "      <td>0</td>\n",
       "      <td>-41.650751</td>\n",
       "      <td>-6.007241</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786073</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137275000</td>\n",
       "      <td>137276000</td>\n",
       "      <td>296</td>\n",
       "      <td>1.990841</td>\n",
       "      <td>268.077912</td>\n",
       "      <td>196.708291</td>\n",
       "      <td>4.575436</td>\n",
       "      <td>229.442056</td>\n",
       "      <td>211.609361</td>\n",
       "      <td>4.655896</td>\n",
       "      <td>228.354529</td>\n",
       "      <td>211.091617</td>\n",
       "      <td>1.921453</td>\n",
       "      <td>268.932391</td>\n",
       "      <td>198.445131</td>\n",
       "      <td>1.881115</td>\n",
       "      <td>269.472162</td>\n",
       "      <td>198.871703</td>\n",
       "      <td>1.974808</td>\n",
       "      <td>268.272874</td>\n",
       "      <td>197.133316</td>\n",
       "      <td>1.977276</td>\n",
       "      <td>268.231681</td>\n",
       "      <td>197.226096</td>\n",
       "      <td>712.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>(True, False)</td>\n",
       "      <td>2.687348</td>\n",
       "      <td>0</td>\n",
       "      <td>27.922088</td>\n",
       "      <td>-0.854479</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786074</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137282000</td>\n",
       "      <td>137283000</td>\n",
       "      <td>211</td>\n",
       "      <td>-1.612483</td>\n",
       "      <td>232.654005</td>\n",
       "      <td>180.337479</td>\n",
       "      <td>-0.329061</td>\n",
       "      <td>215.647526</td>\n",
       "      <td>199.475805</td>\n",
       "      <td>-0.272968</td>\n",
       "      <td>214.851260</td>\n",
       "      <td>199.058885</td>\n",
       "      <td>-1.919711</td>\n",
       "      <td>237.194349</td>\n",
       "      <td>186.184553</td>\n",
       "      <td>-1.941497</td>\n",
       "      <td>237.514975</td>\n",
       "      <td>186.512986</td>\n",
       "      <td>-1.783906</td>\n",
       "      <td>235.171685</td>\n",
       "      <td>183.598811</td>\n",
       "      <td>-1.783008</td>\n",
       "      <td>235.161871</td>\n",
       "      <td>183.634632</td>\n",
       "      <td>572.0</td>\n",
       "      <td>641.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(True, True)</td>\n",
       "      <td>-0.090791</td>\n",
       "      <td>0</td>\n",
       "      <td>-21.654005</td>\n",
       "      <td>-4.540344</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786075</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137290000</td>\n",
       "      <td>137291000</td>\n",
       "      <td>224</td>\n",
       "      <td>-0.215971</td>\n",
       "      <td>226.884483</td>\n",
       "      <td>178.380280</td>\n",
       "      <td>0.419311</td>\n",
       "      <td>218.046575</td>\n",
       "      <td>201.585989</td>\n",
       "      <td>0.479479</td>\n",
       "      <td>217.199655</td>\n",
       "      <td>201.151534</td>\n",
       "      <td>-0.682673</td>\n",
       "      <td>233.338243</td>\n",
       "      <td>187.113535</td>\n",
       "      <td>-0.708182</td>\n",
       "      <td>233.697227</td>\n",
       "      <td>187.501923</td>\n",
       "      <td>-0.522518</td>\n",
       "      <td>231.079535</td>\n",
       "      <td>183.571865</td>\n",
       "      <td>-0.521048</td>\n",
       "      <td>231.060300</td>\n",
       "      <td>183.607554</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>(False, False)</td>\n",
       "      <td>2.162811</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.884483</td>\n",
       "      <td>-6.453759</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786076</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137294000</td>\n",
       "      <td>137295000</td>\n",
       "      <td>182</td>\n",
       "      <td>-5.354983</td>\n",
       "      <td>255.511109</td>\n",
       "      <td>188.447258</td>\n",
       "      <td>-2.785949</td>\n",
       "      <td>221.885053</td>\n",
       "      <td>204.962282</td>\n",
       "      <td>-2.724207</td>\n",
       "      <td>220.957086</td>\n",
       "      <td>204.499773</td>\n",
       "      <td>-5.749158</td>\n",
       "      <td>262.016662</td>\n",
       "      <td>193.710168</td>\n",
       "      <td>-5.780195</td>\n",
       "      <td>262.544150</td>\n",
       "      <td>194.170421</td>\n",
       "      <td>-5.529565</td>\n",
       "      <td>258.109628</td>\n",
       "      <td>189.451152</td>\n",
       "      <td>-5.526926</td>\n",
       "      <td>258.087263</td>\n",
       "      <td>189.520676</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(True, True)</td>\n",
       "      <td>4.476550</td>\n",
       "      <td>0</td>\n",
       "      <td>-73.511109</td>\n",
       "      <td>-6.505553</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1786077 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome      start        end  N_observed  N_bar_7_noncoding  \\\n",
       "0             chr1    1432000    1433000         260          -0.372778   \n",
       "1             chr1    1435000    1436000         275          -6.280673   \n",
       "2             chr1    1449000    1450000         225           0.139977   \n",
       "3             chr1    1450000    1451000         313           2.524948   \n",
       "4             chr1    1451000    1452000         293           2.951595   \n",
       "...            ...        ...        ...         ...                ...   \n",
       "1786072       chr9  137269000  137270000         210          -3.060479   \n",
       "1786073       chr9  137275000  137276000         296           1.990841   \n",
       "1786074       chr9  137282000  137283000         211          -1.612483   \n",
       "1786075       chr9  137290000  137291000         224          -0.215971   \n",
       "1786076       chr9  137294000  137295000         182          -5.354983   \n",
       "\n",
       "         N_mean_null_7_noncoding  N_variance_null_7_noncoding  \\\n",
       "0                     265.321432                   203.777753   \n",
       "1                     368.627875                   222.227572   \n",
       "2                     223.137143                   177.111728   \n",
       "3                     277.859484                   193.691881   \n",
       "4                     252.829443                   185.226049   \n",
       "...                          ...                          ...   \n",
       "1786072               251.650751                   185.210985   \n",
       "1786073               268.077912                   196.708291   \n",
       "1786074               232.654005                   180.337479   \n",
       "1786075               226.884483                   178.380280   \n",
       "1786076               255.511109                   188.447258   \n",
       "\n",
       "         N_bar_1_noncoding  N_mean_null_1_noncoding  \\\n",
       "0                 1.590921               236.519249   \n",
       "1                 2.097960               243.596443   \n",
       "2                 0.959620               211.569144   \n",
       "3                 5.870332               227.882674   \n",
       "4                 4.803930               223.924244   \n",
       "...                    ...                      ...   \n",
       "1786072          -0.657779               219.366052   \n",
       "1786073           4.575436               229.442056   \n",
       "1786074          -0.329061               215.647526   \n",
       "1786075           0.419311               218.046575   \n",
       "1786076          -2.785949               221.885053   \n",
       "\n",
       "         N_variance_null_1_noncoding  N_bar_1_noncoding_noisy  \\\n",
       "0                         217.834403                 1.676924   \n",
       "1                         224.059444                 2.193623   \n",
       "2                         195.888493                 1.011359   \n",
       "3                         210.237742                 5.950235   \n",
       "4                         206.755939                 4.877094   \n",
       "...                              ...                      ...   \n",
       "1786072                   202.746590                -0.596997   \n",
       "1786073                   211.609361                 4.655896   \n",
       "1786074                   199.475805                -0.272968   \n",
       "1786075                   201.585989                 0.479479   \n",
       "1786076                   204.962282                -2.724207   \n",
       "\n",
       "         N_mean_null_1_noncoding_noisy  N_variance_null_1_noncoding_noisy  \\\n",
       "0                           235.282294                         217.264932   \n",
       "1                           242.210058                         223.438246   \n",
       "2                           210.858989                         195.501382   \n",
       "3                           226.828073                         209.731395   \n",
       "4                           222.953222                         206.278524   \n",
       "...                                ...                                ...   \n",
       "1786072                     218.491272                         202.302491   \n",
       "1786073                     228.354529                         211.091617   \n",
       "1786074                     214.851260                         199.058885   \n",
       "1786075                     217.199655                         201.151534   \n",
       "1786076                     220.957086                         204.499773   \n",
       "\n",
       "         N_bar_3_noncoding  N_mean_null_3_noncoding  \\\n",
       "0                -0.247615               263.528599   \n",
       "1                -7.547130               387.939429   \n",
       "2                 0.258359               221.559138   \n",
       "3                 2.197534               281.968822   \n",
       "4                 2.367303               260.131776   \n",
       "...                    ...                      ...   \n",
       "1786072          -3.445370               257.657992   \n",
       "1786073           1.921453               268.932391   \n",
       "1786074          -1.919711               237.194349   \n",
       "1786075          -0.682673               233.338243   \n",
       "1786076          -5.749158               262.016662   \n",
       "\n",
       "         N_variance_null_3_noncoding  N_bar_3_noncoding_noisy  \\\n",
       "0                         203.072610                -0.283079   \n",
       "1                         223.937877                -7.609011   \n",
       "2                         177.372370                 0.225283   \n",
       "3                         199.400099                 2.147648   \n",
       "4                         192.772432                 2.332657   \n",
       "...                              ...                      ...   \n",
       "1786072                   191.337298                -3.474260   \n",
       "1786073                   198.445131                 1.881115   \n",
       "1786074                   186.184553                -1.941497   \n",
       "1786075                   187.113535                -0.708182   \n",
       "1786076                   193.710168                -5.780195   \n",
       "\n",
       "         N_mean_null_3_noncoding_noisy  N_variance_null_3_noncoding_noisy  \\\n",
       "0                           264.038617                         203.540686   \n",
       "1                           389.032484                         224.595337   \n",
       "2                           221.996472                         177.748203   \n",
       "3                           282.634862                         199.905291   \n",
       "4                           260.580640                         193.155585   \n",
       "...                                ...                                ...   \n",
       "1786072                     258.105694                         191.720447   \n",
       "1786073                     269.472162                         198.871703   \n",
       "1786074                     237.514975                         186.512986   \n",
       "1786075                     233.697227                         187.501923   \n",
       "1786076                     262.544150                         194.170421   \n",
       "\n",
       "         N_bar_5_noncoding  N_mean_null_5_noncoding  \\\n",
       "0                -0.339806               264.852332   \n",
       "1                -6.810107               375.796139   \n",
       "2                 0.330975               220.612316   \n",
       "3                 2.480911               278.322530   \n",
       "4                 2.725116               255.650992   \n",
       "...                    ...                      ...   \n",
       "1786072          -3.228932               254.283943   \n",
       "1786073           1.974808               268.272874   \n",
       "1786074          -1.783906               235.171685   \n",
       "1786075          -0.522518               231.079535   \n",
       "1786076          -5.529565               258.109628   \n",
       "\n",
       "         N_variance_null_5_noncoding  N_bar_5_noncoding_noisy  \\\n",
       "0                         203.910737                -0.340167   \n",
       "1                         219.068492                -6.784165   \n",
       "2                         175.743926                 0.330023   \n",
       "3                         195.376527                 2.482902   \n",
       "4                         187.839972                 2.724052   \n",
       "...                              ...                      ...   \n",
       "1786072                   188.093965                -3.225423   \n",
       "1786073                   197.133316                 1.977276   \n",
       "1786074                   183.598811                -1.783008   \n",
       "1786075                   183.571865                -0.521048   \n",
       "1786076                   189.451152                -5.526926   \n",
       "\n",
       "         N_mean_null_5_noncoding_noisy  N_variance_null_5_noncoding_noisy  \\\n",
       "0                           264.858613                         204.004160   \n",
       "1                           375.458319                         219.269877   \n",
       "2                           220.624461                         175.781765   \n",
       "3                           278.288535                         195.445943   \n",
       "4                           255.656435                         187.931976   \n",
       "...                                ...                                ...   \n",
       "1786072                     254.247196                         188.190706   \n",
       "1786073                     268.231681                         197.226096   \n",
       "1786074                     235.161871                         183.634632   \n",
       "1786075                     231.060300                         183.607554   \n",
       "1786076                     258.087263                         189.520676   \n",
       "\n",
       "         enhancer overlap  merged_exon overlap  window overlaps enhancer  \\\n",
       "0                  1000.0                  NaN                      True   \n",
       "1                  1000.0                444.0                      True   \n",
       "2                     NaN                311.0                     False   \n",
       "3                     NaN                758.0                     False   \n",
       "4                     NaN                  NaN                     False   \n",
       "...                   ...                  ...                       ...   \n",
       "1786072            1000.0                  NaN                      True   \n",
       "1786073             712.0                  NaN                      True   \n",
       "1786074             572.0                641.0                      True   \n",
       "1786075               NaN                  NaN                     False   \n",
       "1786076            1000.0               1000.0                      True   \n",
       "\n",
       "         window overlaps merged_exon window overlaps (enhancer, merged_exon)  \\\n",
       "0                              False                           (True, False)   \n",
       "1                               True                            (True, True)   \n",
       "2                               True                           (False, True)   \n",
       "3                               True                           (False, True)   \n",
       "4                              False                          (False, False)   \n",
       "...                              ...                                     ...   \n",
       "1786072                        False                           (True, False)   \n",
       "1786073                        False                           (True, False)   \n",
       "1786074                         True                            (True, True)   \n",
       "1786075                        False                          (False, False)   \n",
       "1786076                         True                            (True, True)   \n",
       "\n",
       "         new chen zscore  positive-vista-enhancer count   delta_sf    delta_f  \\\n",
       "0               4.299894                              0  -5.321432   1.792832   \n",
       "1               3.331645                              0 -93.627875 -19.311554   \n",
       "2               1.817267                              0   1.862857   1.578005   \n",
       "3               1.365863                              0  35.140516  -4.109338   \n",
       "4               0.666316                              0  40.170557  -7.302333   \n",
       "...                  ...                            ...        ...        ...   \n",
       "1786072         5.276351                              0 -41.650751  -6.007241   \n",
       "1786073         2.687348                              0  27.922088  -0.854479   \n",
       "1786074        -0.090791                              0 -21.654005  -4.540344   \n",
       "1786075         2.162811                              0  -2.884483  -6.453759   \n",
       "1786076         4.476550                              0 -73.511109  -6.505553   \n",
       "\n",
       "         window overlaps positive vista enhancer  \n",
       "0                                          False  \n",
       "1                                          False  \n",
       "2                                          False  \n",
       "3                                          False  \n",
       "4                                          False  \n",
       "...                                          ...  \n",
       "1786072                                    False  \n",
       "1786073                                    False  \n",
       "1786074                                    False  \n",
       "1786075                                    False  \n",
       "1786076                                    False  \n",
       "\n",
       "[1786077 rows x 35 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "\n",
    "pd.set_option('display.max_columns', 50)\n",
    "\n",
    "TRAIN_SET_LABEL = 'noncoding'\n",
    "\n",
    "def get_windows_scores_annotations():\n",
    "    filename = f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/mchale.kmerSizes.trainSets.noisy.enhancer-exon.positive-vista-enhancers.bed'\n",
    "    df = pd.read_csv(filename, sep='\\t')\n",
    "    df['delta_sf'] = df['N_observed'] - df[f'N_mean_null_7_{TRAIN_SET_LABEL}']\n",
    "    df['delta_f'] = df[f'N_mean_null_7_{TRAIN_SET_LABEL}'] - df[f'N_mean_null_3_{TRAIN_SET_LABEL}']\n",
    "    df['window overlaps positive vista enhancer'] = df['positive-vista-enhancer count'] > 0    \n",
    "    return df\n",
    "\n",
    "windows_scores_annotations = get_windows_scores_annotations()\n",
    "windows_scores_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c895154a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "windows_scores_annotations_coding = windows_scores_annotations[windows_scores_annotations['window overlaps merged_exon'] == True]\n",
    "windows_scores_annotations_noncoding = windows_scores_annotations[windows_scores_annotations['window overlaps merged_exon'] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fbd82950",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns \n",
    "sns.set(font_scale = 1.5)\n",
    "\n",
    "import matplotlib.pyplot as plt \n",
    "plt.rcParams.update({\n",
    "    'font.size': 20,\n",
    "    'figure.figsize': (18, 10)\n",
    "})\n",
    "\n",
    "import numpy as np "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b6b896e",
   "metadata": {},
   "source": [
    "## Compute $w_{s, exact}$ , $w_{s, approx}$ , and $n_s$ for a subset of windows, and fetch the windows' sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7c2e91d4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 17.39it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>N_observed</th>\n",
       "      <th>N_bar_7_noncoding</th>\n",
       "      <th>N_mean_null_7_noncoding</th>\n",
       "      <th>N_variance_null_7_noncoding</th>\n",
       "      <th>N_bar_1_noncoding</th>\n",
       "      <th>N_mean_null_1_noncoding</th>\n",
       "      <th>N_variance_null_1_noncoding</th>\n",
       "      <th>N_bar_1_noncoding_noisy</th>\n",
       "      <th>N_mean_null_1_noncoding_noisy</th>\n",
       "      <th>N_variance_null_1_noncoding_noisy</th>\n",
       "      <th>N_bar_3_noncoding</th>\n",
       "      <th>N_mean_null_3_noncoding</th>\n",
       "      <th>N_variance_null_3_noncoding</th>\n",
       "      <th>N_bar_3_noncoding_noisy</th>\n",
       "      <th>N_mean_null_3_noncoding_noisy</th>\n",
       "      <th>N_variance_null_3_noncoding_noisy</th>\n",
       "      <th>N_bar_5_noncoding</th>\n",
       "      <th>N_mean_null_5_noncoding</th>\n",
       "      <th>N_variance_null_5_noncoding</th>\n",
       "      <th>N_bar_5_noncoding_noisy</th>\n",
       "      <th>N_mean_null_5_noncoding_noisy</th>\n",
       "      <th>N_variance_null_5_noncoding_noisy</th>\n",
       "      <th>enhancer overlap</th>\n",
       "      <th>merged_exon overlap</th>\n",
       "      <th>window overlaps enhancer</th>\n",
       "      <th>window overlaps merged_exon</th>\n",
       "      <th>window overlaps (enhancer, merged_exon)</th>\n",
       "      <th>new chen zscore</th>\n",
       "      <th>positive-vista-enhancer count</th>\n",
       "      <th>delta_sf</th>\n",
       "      <th>delta_f</th>\n",
       "      <th>window overlaps positive vista enhancer</th>\n",
       "      <th>n_marginal_local</th>\n",
       "      <th>w_exact</th>\n",
       "      <th>w_approx</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1432000</td>\n",
       "      <td>1433000</td>\n",
       "      <td>260</td>\n",
       "      <td>-0.372778</td>\n",
       "      <td>265.321432</td>\n",
       "      <td>203.777753</td>\n",
       "      <td>1.590921</td>\n",
       "      <td>236.519249</td>\n",
       "      <td>217.834403</td>\n",
       "      <td>1.676924</td>\n",
       "      <td>235.282294</td>\n",
       "      <td>217.264932</td>\n",
       "      <td>-0.247615</td>\n",
       "      <td>263.528599</td>\n",
       "      <td>203.072610</td>\n",
       "      <td>-0.283079</td>\n",
       "      <td>264.038617</td>\n",
       "      <td>203.540686</td>\n",
       "      <td>-0.339806</td>\n",
       "      <td>264.852332</td>\n",
       "      <td>203.910737</td>\n",
       "      <td>-0.340167</td>\n",
       "      <td>264.858613</td>\n",
       "      <td>204.004160</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>(True, False)</td>\n",
       "      <td>4.299894</td>\n",
       "      <td>0</td>\n",
       "      <td>-5.321432</td>\n",
       "      <td>1.792832</td>\n",
       "      <td>False</td>\n",
       "      <td>{'AAA': 6, 'AAC'...</td>\n",
       "      <td>{'AAA': 0.088171...</td>\n",
       "      <td>{'AAA': 0.109115...</td>\n",
       "      <td>GTGGGCTGGTCTGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1435000</td>\n",
       "      <td>1436000</td>\n",
       "      <td>275</td>\n",
       "      <td>-6.280673</td>\n",
       "      <td>368.627875</td>\n",
       "      <td>222.227572</td>\n",
       "      <td>2.097960</td>\n",
       "      <td>243.596443</td>\n",
       "      <td>224.059444</td>\n",
       "      <td>2.193623</td>\n",
       "      <td>242.210058</td>\n",
       "      <td>223.438246</td>\n",
       "      <td>-7.547130</td>\n",
       "      <td>387.939429</td>\n",
       "      <td>223.937877</td>\n",
       "      <td>-7.609011</td>\n",
       "      <td>389.032484</td>\n",
       "      <td>224.595337</td>\n",
       "      <td>-6.810107</td>\n",
       "      <td>375.796139</td>\n",
       "      <td>219.068492</td>\n",
       "      <td>-6.784165</td>\n",
       "      <td>375.458319</td>\n",
       "      <td>219.269877</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>444.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(True, True)</td>\n",
       "      <td>3.331645</td>\n",
       "      <td>0</td>\n",
       "      <td>-93.627875</td>\n",
       "      <td>-19.311554</td>\n",
       "      <td>False</td>\n",
       "      <td>{'AAA': 1, 'AAC'...</td>\n",
       "      <td>{'AAA': 0.065959...</td>\n",
       "      <td>{'AAA': 0.109115...</td>\n",
       "      <td>AGGCCCAGTGTGGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1449000</td>\n",
       "      <td>1450000</td>\n",
       "      <td>225</td>\n",
       "      <td>0.139977</td>\n",
       "      <td>223.137143</td>\n",
       "      <td>177.111728</td>\n",
       "      <td>0.959620</td>\n",
       "      <td>211.569144</td>\n",
       "      <td>195.888493</td>\n",
       "      <td>1.011359</td>\n",
       "      <td>210.858989</td>\n",
       "      <td>195.501382</td>\n",
       "      <td>0.258359</td>\n",
       "      <td>221.559138</td>\n",
       "      <td>177.372370</td>\n",
       "      <td>0.225283</td>\n",
       "      <td>221.996472</td>\n",
       "      <td>177.748203</td>\n",
       "      <td>0.330975</td>\n",
       "      <td>220.612316</td>\n",
       "      <td>175.743926</td>\n",
       "      <td>0.330023</td>\n",
       "      <td>220.624461</td>\n",
       "      <td>175.781765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>311.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>(False, True)</td>\n",
       "      <td>1.817267</td>\n",
       "      <td>0</td>\n",
       "      <td>1.862857</td>\n",
       "      <td>1.578005</td>\n",
       "      <td>False</td>\n",
       "      <td>{'AAA': 17, 'AAC...</td>\n",
       "      <td>{'AAA': 0.104229...</td>\n",
       "      <td>{'AAA': 0.109115...</td>\n",
       "      <td>cctcctgaactgcccg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome    start      end  N_observed  N_bar_7_noncoding  \\\n",
       "0       chr1  1432000  1433000         260          -0.372778   \n",
       "1       chr1  1435000  1436000         275          -6.280673   \n",
       "2       chr1  1449000  1450000         225           0.139977   \n",
       "\n",
       "   N_mean_null_7_noncoding  N_variance_null_7_noncoding  N_bar_1_noncoding  \\\n",
       "0           265.321432               203.777753                   1.590921   \n",
       "1           368.627875               222.227572                   2.097960   \n",
       "2           223.137143               177.111728                   0.959620   \n",
       "\n",
       "   N_mean_null_1_noncoding  N_variance_null_1_noncoding  \\\n",
       "0           236.519249               217.834403           \n",
       "1           243.596443               224.059444           \n",
       "2           211.569144               195.888493           \n",
       "\n",
       "   N_bar_1_noncoding_noisy  N_mean_null_1_noncoding_noisy  \\\n",
       "0             1.676924               235.282294             \n",
       "1             2.193623               242.210058             \n",
       "2             1.011359               210.858989             \n",
       "\n",
       "   N_variance_null_1_noncoding_noisy  N_bar_3_noncoding  \\\n",
       "0           217.264932                        -0.247615   \n",
       "1           223.438246                        -7.547130   \n",
       "2           195.501382                         0.258359   \n",
       "\n",
       "   N_mean_null_3_noncoding  N_variance_null_3_noncoding  \\\n",
       "0           263.528599               203.072610           \n",
       "1           387.939429               223.937877           \n",
       "2           221.559138               177.372370           \n",
       "\n",
       "   N_bar_3_noncoding_noisy  N_mean_null_3_noncoding_noisy  \\\n",
       "0            -0.283079               264.038617             \n",
       "1            -7.609011               389.032484             \n",
       "2             0.225283               221.996472             \n",
       "\n",
       "   N_variance_null_3_noncoding_noisy  N_bar_5_noncoding  \\\n",
       "0           203.540686                        -0.339806   \n",
       "1           224.595337                        -6.810107   \n",
       "2           177.748203                         0.330975   \n",
       "\n",
       "   N_mean_null_5_noncoding  N_variance_null_5_noncoding  \\\n",
       "0           264.852332               203.910737           \n",
       "1           375.796139               219.068492           \n",
       "2           220.612316               175.743926           \n",
       "\n",
       "   N_bar_5_noncoding_noisy  N_mean_null_5_noncoding_noisy  \\\n",
       "0            -0.340167               264.858613             \n",
       "1            -6.784165               375.458319             \n",
       "2             0.330023               220.624461             \n",
       "\n",
       "   N_variance_null_5_noncoding_noisy  enhancer overlap  merged_exon overlap  \\\n",
       "0           204.004160                          1000.0                  NaN   \n",
       "1           219.269877                          1000.0                444.0   \n",
       "2           175.781765                             NaN                311.0   \n",
       "\n",
       "   window overlaps enhancer  window overlaps merged_exon  \\\n",
       "0                 True                     False           \n",
       "1                 True                      True           \n",
       "2                False                      True           \n",
       "\n",
       "  window overlaps (enhancer, merged_exon)  new chen zscore  \\\n",
       "0        (True, False)                            4.299894   \n",
       "1         (True, True)                            3.331645   \n",
       "2        (False, True)                            1.817267   \n",
       "\n",
       "   positive-vista-enhancer count   delta_sf    delta_f  \\\n",
       "0                    0            -5.321432   1.792832   \n",
       "1                    0           -93.627875 -19.311554   \n",
       "2                    0             1.862857   1.578005   \n",
       "\n",
       "   window overlaps positive vista enhancer     n_marginal_local  \\\n",
       "0                False                      {'AAA': 6, 'AAC'...   \n",
       "1                False                      {'AAA': 1, 'AAC'...   \n",
       "2                False                      {'AAA': 17, 'AAC...   \n",
       "\n",
       "               w_exact             w_approx             sequence  \n",
       "0  {'AAA': 0.088171...  {'AAA': 0.109115...  GTGGGCTGGTCTGAGG...  \n",
       "1  {'AAA': 0.065959...  {'AAA': 0.109115...  AGGCCCAGTGTGGTCC...  \n",
       "2  {'AAA': 0.104229...  {'AAA': 0.109115...  cctcctgaactgcccg...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append(f'{CONSTRAINT_TOOLS}/utilities')\n",
    "\n",
    "from collections import defaultdict \n",
    "\n",
    "from read_model import read_model \n",
    "from windows import compute_flanks\n",
    "from get_p0s_p1s_p2s_p3s import get_substitution_probability\n",
    "from kmer import compute_kmers \n",
    "\n",
    "def initialize_nested_dictionary(value): \n",
    "    return {\n",
    "        s: {\n",
    "            f: value for f in compute_kmers(kmer_size=4)\n",
    "        } for s in compute_kmers(kmer_size=3)\n",
    "    }\n",
    "\n",
    "def initialize_dictionary(value): \n",
    "    return {s: value for s in compute_kmers(kmer_size=3)}\n",
    "\n",
    "def read_model_wrapper(kmer_size):\n",
    "    filename = f'{CONSTRAINT_TOOLS}/dist/model-germline-grch38-Nonly.kmerSize-{kmer_size}.trainSet-{TRAIN_SET_LABEL}.json'\n",
    "    model = read_model(filename)\n",
    "    return model\n",
    "\n",
    "def compute_left_right(): \n",
    "    left_flank, right_flank = compute_flanks(3)\n",
    "    left = 3 - left_flank # middle index of 7mer is 3\n",
    "    right = 3 + right_flank + 1 # middle index of 7mer is 3\n",
    "    return left, right\n",
    "\n",
    "def decompose(seven_mer):\n",
    "    left, right = compute_left_right()\n",
    "    s = seven_mer[left:right]\n",
    "    f = seven_mer[:left] + seven_mer[right:]\n",
    "    return s, f \n",
    "\n",
    "def compute_expected_number_ALTs(model_7): \n",
    "    k = initialize_nested_dictionary(value=None)\n",
    "    for seven_mer in model_7['kmerCounts'].keys(): \n",
    "        s, f = decompose(seven_mer)\n",
    "        p1, p2, p3 = (\n",
    "            get_substitution_probability(model_7, seven_mer, ALT_multiplicity) \n",
    "            for ALT_multiplicity in [1, 2, 3]\n",
    "        )\n",
    "        k[s][f] = 1*p1 + 2*p2 + 3*p3\n",
    "    return k\n",
    "\n",
    "def compute_marginal_counts(n): \n",
    "    n_marginal = initialize_dictionary(value=0)\n",
    "    for s in n.keys(): \n",
    "        n_marginal[s] = sum(n[s].values())\n",
    "    return n_marginal \n",
    "\n",
    "def compute_sf_counts_global(model_7): \n",
    "    n = initialize_nested_dictionary(value=0)\n",
    "    for seven_mer in model_7['kmerCounts'].keys():\n",
    "        s, f = decompose(seven_mer)\n",
    "        n[s][f] = model_7['kmerCounts'][seven_mer]['count']\n",
    "    \n",
    "    n_marginal = compute_marginal_counts(n)\n",
    "    \n",
    "    return n, n_marginal\n",
    "\n",
    "import pysam \n",
    "from tqdm import tqdm \n",
    "\n",
    "tqdm.pandas()\n",
    "\n",
    "from kmer import fetch_kmers \n",
    "from pack_unpack import pack\n",
    "\n",
    "def compute_sf_counts_local(chromosome, start, end, genome): \n",
    "    region = pack(chromosome, start, end)\n",
    "    n = initialize_nested_dictionary(value=0)\n",
    "    for seven_mer in fetch_kmers(region, genome, kmer_size=7, log=False): \n",
    "        s, f = decompose(seven_mer)\n",
    "        n[s][f] += 1\n",
    "    \n",
    "    n_marginal = compute_marginal_counts(n)\n",
    "    \n",
    "    return n, n_marginal\n",
    "\n",
    "def compute_cond_f_dist(n, n_marginal):\n",
    "    p = initialize_nested_dictionary(value=None)\n",
    "    for s in n.keys():\n",
    "        if n_marginal[s] > 0: \n",
    "            for f in n[s].keys():\n",
    "                p[s][f] = n[s][f]/n_marginal[s]    \n",
    "    return p\n",
    "\n",
    "def compute_w(expected_number_ALTs, f_cond_dist, n_marginal):\n",
    "    w = initialize_dictionary(value=None)\n",
    "    for s in n_marginal.keys(): \n",
    "        if n_marginal[s] > 0: \n",
    "            w[s] = sum(expected_number_ALTs[s][f] * f_cond_dist[s][f] for f in expected_number_ALTs[s].keys())    \n",
    "    return w \n",
    "\n",
    "def compute_expected_number_ALT(s, model_3): # for sanity check\n",
    "    p1, p2, p3 = (\n",
    "        get_substitution_probability(model_3, s, ALT_multiplicity)\n",
    "        for ALT_multiplicity in [1, 2, 3]\n",
    "    )\n",
    "    return 1*p1 + 2*p2 + 3*p3\n",
    "\n",
    "def sanity_check_1(sub_probs, f_cond_dist_global, n_marginal_global): \n",
    "    from kmer import CpG\n",
    "    # check the formula relating k_s to k_sf, where k is expected number of ALTs\n",
    "    model_3 = read_model_wrapper(kmer_size=3)\n",
    "    w_approx = compute_w(sub_probs, f_cond_dist_global, n_marginal_global)\n",
    "    predictions, actuals = [], []\n",
    "    for s in sub_probs.keys():\n",
    "        prediction = w_approx[s] # predicted expected number of ALTs\n",
    "        actual = compute_expected_number_ALT(s, model_3) # actual expected number of ALTs\n",
    "        if actual > 0.8: print(s, CpG(s), actual)\n",
    "        predictions.append(prediction) \n",
    "        actuals.append(actual)\n",
    "    plt.plot(predictions, actuals, 'o')\n",
    "    plt.plot([0, 1], [0, 1])\n",
    "    \n",
    "def compute_n_predicted(w, n_marginal_local): # for sanity check\n",
    "    n_predicted = 0.0\n",
    "    for s in n_marginal_local.keys(): \n",
    "        if n_marginal_local[s] > 0: \n",
    "             n_predicted += n_marginal_local[s]*w[s]\n",
    "    return n_predicted\n",
    "    \n",
    "def sanity_check_2(n_exact, n_approx, w_exact, w_approx, n_marginal_local): \n",
    "    # n_exact comes from df[f'N_mean_null_7_{TRAIN_SET_LABEL}']\n",
    "    # n_approx comes from df[f'N_mean_null_3_{TRAIN_SET_LABEL}']\n",
    "    n_exact_predicted = compute_n_predicted(w_exact, n_marginal_local)\n",
    "    print('exact', n_exact, n_exact_predicted)\n",
    "    n_approx_predicted = compute_n_predicted(w_approx, n_marginal_local)\n",
    "    print('approx', n_approx, n_approx_predicted) \n",
    "    \n",
    "def add_w_and_nMarginal_row(row, genome, expected_number_ALTs, w_approx): \n",
    "    n_local, n_marginal_local = compute_sf_counts_local(\n",
    "        row['chromosome'], \n",
    "        row['start'], \n",
    "        row['end'], \n",
    "        genome\n",
    "    )\n",
    "    f_cond_dist_local = compute_cond_f_dist(n_local, n_marginal_local)\n",
    "    row['n_marginal_local'] = n_marginal_local\n",
    "    row['w_exact'] = compute_w(expected_number_ALTs, f_cond_dist_local, n_marginal_local)\n",
    "    row['w_approx'] = w_approx\n",
    "    return row\n",
    "\n",
    "def add_w_and_nMarginal(df, genome_filename): \n",
    "    model_7 = read_model_wrapper(kmer_size=7)\n",
    "    expected_number_ALTs = compute_expected_number_ALTs(model_7)\n",
    "    n_global, n_marginal_global = compute_sf_counts_global(model_7)\n",
    "    f_cond_dist_global = compute_cond_f_dist(n_global, n_marginal_global)\n",
    "    # return sanity_check_1(expected_number_ALTs, f_cond_dist_global, n_marginal_global)\n",
    "    w_approx = compute_w(expected_number_ALTs, f_cond_dist_global, n_marginal_global)\n",
    "    \n",
    "    with pysam.FastaFile(genome_filename) as genome: \n",
    "        df = df.progress_apply(\n",
    "            lambda row: add_w_and_nMarginal_row(\n",
    "                row, genome, expected_number_ALTs, w_approx\n",
    "            ), \n",
    "            axis=1\n",
    "        )\n",
    "\n",
    "    return df\n",
    "\n",
    "# /uufs/chpc.utah.edu/common/HIPAA/u6018199/.conda/envs/constraint-tools/lib/python3.9/site-packages/pyranges/get_fasta.py\n",
    "# For more pyranges stuff, see: experiments/germline-model/cpg-islands/cpg-islands-compute-zscores.500%2C600.ipynb\n",
    "\n",
    "import pyranges as pr\n",
    "\n",
    "def add_sequence(df, genome_filename):\n",
    "    columns = ['chromosome', 'start', 'end']\n",
    "    gr = pr.PyRanges(df[columns].rename(columns = {\n",
    "        column: column.title()\n",
    "        for column in columns\n",
    "    }))\n",
    "    sequence = pr.get_fasta(gr, genome_filename)\n",
    "    df['sequence'] = sequence\n",
    "    return df \n",
    "\n",
    "pd.set_option('max_colwidth', 20)\n",
    "\n",
    "def add_w_and_nMarginal_and_sequence(df):\n",
    "    genome_filename = '/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/reference/grch38/hg38.analysisSet.fa.gz'    \n",
    "    df = add_w_and_nMarginal(df, genome_filename)\n",
    "    df = add_sequence(df, genome_filename)\n",
    "    return df \n",
    "\n",
    "add_w_and_nMarginal_and_sequence(windows_scores_annotations.iloc[:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90cd1164",
   "metadata": {},
   "source": [
    "## [TODO #1.2] Choose most constrained window(s), and plot heat map of w_{s,exact}, heat map of w_{s,approx}, and scatter plot of the first w against the second w by ranging over s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc867d15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "079df4d9",
   "metadata": {},
   "source": [
    "## [TODO #1.3] Cluster windows by delta_w_s = w_exact_s - w_approx_s\n",
    "\n",
    "1. each window has a vector <delta_w_s1, delta_w_s2, ...> \n",
    "2. get 100 windows with low zscores, 100 windows with medium zscores, 100 windows with high zscore \n",
    "3. do PCA on all 300 windows  \n",
    "4. ask whether we find three clusters in pc1-pc2 space!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bba48f2f",
   "metadata": {},
   "source": [
    "## [TODO #1.4] [GENERAL] Cluster windows by n_s \n",
    "\n",
    "1. each window has a vector <n_s1, n_s2, ....> \n",
    "2. get 100 windows with low zscores, 100 windows with medium zscores, 100 windows with high zscore \n",
    "3. do PCA on all 300 windows\n",
    "4. ask whether we find three clusters in pc1-pc2 space!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb4af510",
   "metadata": {},
   "source": [
    "## [TODO #2] [COOL] Dissect the apparently constrained windows using Nucleotide Transformer to make discoveries about genome biology\n",
    "\n",
    "https://github.com/instadeepai/nucleotide-transformer\n",
    "\n",
    "0. map windows to their embedding vectors, and show that windows with low-zscores form one cluster, and those with high zscores form a different cluster (similar to n_s analysis above) \n",
    "1. [INCREASE SPATIAL RESOLUTION] Compute attention weights for low-zscore windows using template code in https://github.com/instadeepai/nucleotide-transformer/blob/efc3fa5b6bc57ae8ec2704c792cfb038dd0c689e/examples/inference.ipynb. Do pairs of 100bp regions with high attention also have low SNV density? Does this happen for windows that don’t harbor known enhancers, potentially discovering new enhancers? For windows that harbor enhancers, does the attention scores coincide with the enhancers? [Aggregate attention scores, weighting them by an enhancer matrix, similar to, e.g., Fig 3C, to identify layers and heads that are paying attention to enhancers in a window that contains the enhancer.]\n",
    "2. Compute predicted effect of SNVs in a window using distance between embeddings of REF and ALT alleles for each site (as done in Fig 4A of nucleotide transformer paper). Average predicted effects over all SNP sites in window. Does average predicted effect correlate with zscore of window? One could also repeat this analysis for all monomorphic sites. (This result would nicely parallel results regarding singleton density.)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533f9c41",
   "metadata": {},
   "source": [
    "## [TODO] Aaron's questions (may help discover new biology)\n",
    "\n",
    "1. Correlate z-scores of windows with average CpG content (relative to genome-wide average) of windows [compare with https://docs.google.com/presentation/d/1XzvnTeJBGnXG9t2UhkrAEBUdemvSNydwC25zglp5Y14/edit#slide=id.g230fb7471c0_0_150 )\n",
    "2. If 1. Succeeds, then do: Fraction of noncoding windows that overlap CpG islands vs fraction of extreme-zscore windows that overlaps CpG islands \n",
    "\n",
    "Are zscores correlated with Ultraconserved elements? \n",
    "Fetch ultra conserved elements from zoonomia paper: https://quinlangroup.slack.com/archives/C449KJT3J/p1683166940410939\n",
    "* [ Aaron’s earlier remarks on UCEs: https://quinlangroup.slack.com/archives/C01U5FGKAHM/p1671468415354519\n",
    "* [ Also, bear in mind prior conservation scores: \n",
    "\t* Supplementary Figure 7 of https://www.nature.com/articles/s41588-018-0062-7 shows that low-SNV-density regions are largely different from high conservation regions; \n",
    "\t* LOGO paper and Venter paper have GERP data links\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4420e0ce",
   "metadata": {},
   "source": [
    "## [TODO] compute delta_f, where f = methylation level, for each window, and recompute the heat map in the delta_sf-delta_f plane, where this time delta_sf is replaced by z_chen\n",
    "\n",
    "Hopefully, this will show that measuring how anomalous the methylation levels are can improve upon the Chen zscore for many surrogates of constraint. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "866daec6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
