{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "CONSTRAINT_TOOLS = '/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools'\n",
    "CONSTRAINT_TOOLS_DATA = '/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools'\n",
    "\n",
    "import sys\n",
    "sys.path.append(f'{CONSTRAINT_TOOLS}/utilities')\n",
    "\n",
    "LOWER_DELETION_SIZE_LIMIT = 4000\n",
    "UPPER_DELETION_SIZE_LIMIT = 10000\n",
    "DELETION_ALLELE_FREQ_THRESHOLD = 0.00001\n",
    "\n",
    "# DELETION_TYPE = 'long' \n",
    "DELETION_TYPE = 'all' \n",
    "\n",
    "# TODO: remove? \n",
    "BALANCE_CLASSES = False\n",
    "\n",
    "PUBLIC_REPO = 'constraint-tools-data'\n",
    "PUBLIC_REPO_DIR = f'/scratch/ucgd/lustre-work/quinlan/u6018199/{PUBLIC_REPO}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute SV scores of a set of positive and negative enhancers, i.e., enhancers linked to genes that are known to be under negative selection or no selection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 1. read in enhancers into dataframe \n",
    "# 2. write out a file in which the first three columns are the enhancer coordinates\n",
    "# 3. assign that file to the \"windows\" variable below, and edit the following code to compute SV scores for each enhancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>enhancer</th>\n",
       "      <th>enhancer_Gnocchi</th>\n",
       "      <th>Haploinsufficient</th>\n",
       "      <th>MGI essential</th>\n",
       "      <th>OMIM dominant</th>\n",
       "      <th>LOEUF constrained</th>\n",
       "      <th>Olfactory</th>\n",
       "      <th>LOEUF unconstrained</th>\n",
       "      <th>LOEUF underpowered</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>chr1-867020-867220</td>\n",
       "      <td>1.509945</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1-923220-923420</td>\n",
       "      <td>1.579239</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RNF223</td>\n",
       "      <td>chr1-1094220-1094420</td>\n",
       "      <td>2.725941</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HES4</td>\n",
       "      <td>chr1-1031020-1031220</td>\n",
       "      <td>4.830528</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGRN</td>\n",
       "      <td>chr1-1003620-1003820</td>\n",
       "      <td>5.938154</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15536</th>\n",
       "      <td>PNPLA7</td>\n",
       "      <td>chr9-137486748-137487348</td>\n",
       "      <td>6.012245</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15537</th>\n",
       "      <td>ZMYND19</td>\n",
       "      <td>chr9-137532548-137532748</td>\n",
       "      <td>2.448615</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15538</th>\n",
       "      <td>ARRDC1</td>\n",
       "      <td>chr9-137593948-137594348</td>\n",
       "      <td>1.190685</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15539</th>\n",
       "      <td>EHMT1</td>\n",
       "      <td>chr9-137649348-137649548</td>\n",
       "      <td>3.038869</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15540</th>\n",
       "      <td>CACNA1B</td>\n",
       "      <td>chr9-137899948-137900348</td>\n",
       "      <td>4.984796</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15541 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                  enhancer  enhancer_Gnocchi  Haploinsufficient  \\\n",
       "0        NOC2L        chr1-867020-867220          1.509945              False   \n",
       "1       SAMD11        chr1-923220-923420          1.579239              False   \n",
       "2       RNF223      chr1-1094220-1094420          2.725941              False   \n",
       "3         HES4      chr1-1031020-1031220          4.830528              False   \n",
       "4         AGRN      chr1-1003620-1003820          5.938154              False   \n",
       "...        ...                       ...               ...                ...   \n",
       "15536   PNPLA7  chr9-137486748-137487348          6.012245              False   \n",
       "15537  ZMYND19  chr9-137532548-137532748          2.448615              False   \n",
       "15538   ARRDC1  chr9-137593948-137594348          1.190685              False   \n",
       "15539    EHMT1  chr9-137649348-137649548          3.038869               True   \n",
       "15540  CACNA1B  chr9-137899948-137900348          4.984796              False   \n",
       "\n",
       "       MGI essential  OMIM dominant  LOEUF constrained  Olfactory  \\\n",
       "0              False          False              False      False   \n",
       "1              False          False              False      False   \n",
       "2              False          False              False      False   \n",
       "3              False          False              False      False   \n",
       "4               True          False              False      False   \n",
       "...              ...            ...                ...        ...   \n",
       "15536          False          False              False      False   \n",
       "15537          False          False              False      False   \n",
       "15538          False          False              False      False   \n",
       "15539           True           True               True      False   \n",
       "15540           True          False               True      False   \n",
       "\n",
       "       LOEUF unconstrained  LOEUF underpowered  \n",
       "0                    False               False  \n",
       "1                    False               False  \n",
       "2                    False                True  \n",
       "3                    False                True  \n",
       "4                    False               False  \n",
       "...                    ...                 ...  \n",
       "15536                False               False  \n",
       "15537                False               False  \n",
       "15538                False               False  \n",
       "15539                False               False  \n",
       "15540                False               False  \n",
       "\n",
       "[15541 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "\n",
    "pd.read_csv(\n",
    "  f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_6_ESM.txt', \n",
    "  sep='\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;36mWe assume that the first line of the following is a header line: \u001b[0m/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/genes/grch38/canonical-exons.positive.sorted.bed\n",
      "\u001b[0;36mWrote windows with deletion counts and overlaps to: \u001b[0m/scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/benchmark-genome-wide-predictions/chen-et-al-2022/positive-canonical-exons-intersect-all-observed-deletions.bed\n",
      "\u001b[0;36mWrote all observed deletions in UCSC-genome-browser format to: \u001b[0m/scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools-data/all-observed-deletions.bed\n",
      "Everything up-to-date\n",
      "\u001b[0;36mPushed /scratch/ucgd/lustre-work/quinlan/u6018199/constraint-tools-data/all-observed-deletions.bed to public repo\u001b[0m\n",
      "On branch main\n",
      "Your branch is up to date with 'origin/main'.\n",
      "\n",
      "nothing to commit, working tree clean\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>exon_count</th>\n",
       "      <th>deletion count over all exons</th>\n",
       "      <th>deletion overlap over all exons</th>\n",
       "      <th>total length of all exons</th>\n",
       "      <th>gene coordinates</th>\n",
       "      <th>deletion count per bp</th>\n",
       "      <th>fraction of bps that are deleted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTN2</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>407</td>\n",
       "      <td>4851</td>\n",
       "      <td>chr1:236686499-236764631</td>\n",
       "      <td>0.000618</td>\n",
       "      <td>0.083900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADAR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6595</td>\n",
       "      <td>chr1:154582057-154608186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AHDC1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6431</td>\n",
       "      <td>chr1:27534303-27603632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMPD1</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>866</td>\n",
       "      <td>2359</td>\n",
       "      <td>chr1:114673098-114695586</td>\n",
       "      <td>0.002543</td>\n",
       "      <td>0.367105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APOA2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>470</td>\n",
       "      <td>chr1:161222292-161223628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>TGFBR1</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>869</td>\n",
       "      <td>5924</td>\n",
       "      <td>chr9:99105150-99153658</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.146691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>TOR1A</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2075</td>\n",
       "      <td>chr9:129812942-129824136</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>TPM2</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>1318</td>\n",
       "      <td>1318</td>\n",
       "      <td>chr9:35682926-35690056</td>\n",
       "      <td>0.007587</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "      <td>3792</td>\n",
       "      <td>8593</td>\n",
       "      <td>chr9:132891348-132944633</td>\n",
       "      <td>0.000698</td>\n",
       "      <td>0.441289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>VCP</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3729</td>\n",
       "      <td>chr9:35056064-35072625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>646 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_symbol  exon_count  deletion count over all exons  \\\n",
       "0         ACTN2          21                              3   \n",
       "1          ADAR          15                              0   \n",
       "2         AHDC1           7                              0   \n",
       "3         AMPD1          16                              6   \n",
       "4         APOA2           4                              0   \n",
       "..          ...         ...                            ...   \n",
       "641      TGFBR1           9                              4   \n",
       "642       TOR1A           5                              0   \n",
       "643        TPM2           9                             10   \n",
       "644        TSC1          23                              6   \n",
       "645         VCP          17                              0   \n",
       "\n",
       "     deletion overlap over all exons  total length of all exons  \\\n",
       "0                                407                       4851   \n",
       "1                                  0                       6595   \n",
       "2                                  0                       6431   \n",
       "3                                866                       2359   \n",
       "4                                  0                        470   \n",
       "..                               ...                        ...   \n",
       "641                              869                       5924   \n",
       "642                                0                       2075   \n",
       "643                             1318                       1318   \n",
       "644                             3792                       8593   \n",
       "645                                0                       3729   \n",
       "\n",
       "             gene coordinates  deletion count per bp  \\\n",
       "0    chr1:236686499-236764631               0.000618   \n",
       "1    chr1:154582057-154608186               0.000000   \n",
       "2      chr1:27534303-27603632               0.000000   \n",
       "3    chr1:114673098-114695586               0.002543   \n",
       "4    chr1:161222292-161223628               0.000000   \n",
       "..                        ...                    ...   \n",
       "641    chr9:99105150-99153658               0.000675   \n",
       "642  chr9:129812942-129824136               0.000000   \n",
       "643    chr9:35682926-35690056               0.007587   \n",
       "644  chr9:132891348-132944633               0.000698   \n",
       "645    chr9:35056064-35072625               0.000000   \n",
       "\n",
       "     fraction of bps that are deleted  \n",
       "0                            0.083900  \n",
       "1                            0.000000  \n",
       "2                            0.000000  \n",
       "3                            0.367105  \n",
       "4                            0.000000  \n",
       "..                                ...  \n",
       "641                          0.146691  \n",
       "642                          0.000000  \n",
       "643                          1.000000  \n",
       "644                          0.441289  \n",
       "645                          0.000000  \n",
       "\n",
       "[646 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def assign_SV_scores_to_labeled_genes(\n",
    "    gene_class, \n",
    "    deletion_class,\n",
    "    deletion_type=DELETION_TYPE,\n",
    "):\n",
    "  windows = f'{CONSTRAINT_TOOLS_DATA}/genes/grch38/canonical-exons.{gene_class}.sorted.bed'\n",
    "  windows_with_deletions = f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/{gene_class}-canonical-exons-intersect-{deletion_type}-{deletion_class}-deletions.bed'\n",
    "\n",
    "  intersect_script = f'{CONSTRAINT_TOOLS}/experiments/germline-model/chen-et-al-2022/intersect-windows-with-deletions.sh'\n",
    "\n",
    "  PROCESSED_DELETIONS = f'{deletion_type}-{deletion_class}-deletions.bed'\n",
    "\n",
    "  cmd = (\n",
    "    f'bash '\n",
    "    f'{intersect_script} '\n",
    "    f'{windows} {deletion_class} {windows_with_deletions} '\n",
    "    f'{PUBLIC_REPO_DIR} {PROCESSED_DELETIONS} '\n",
    "    f'{deletion_type} '\n",
    "    f'{LOWER_DELETION_SIZE_LIMIT} {UPPER_DELETION_SIZE_LIMIT} {DELETION_ALLELE_FREQ_THRESHOLD}'\n",
    "  )\n",
    "  print(shell(cmd))\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    windows_with_deletions, \n",
    "    sep='\\t',\n",
    "  )\n",
    "\n",
    "  df = df[\n",
    "    (df['chromosome'] != 'chrX') &\n",
    "    (df['chromosome'] != 'chrY')\n",
    "  ]\n",
    "\n",
    "  # though unlikely, some exons may intersect multiple merged (and therefore non-overlapping) deletions,\n",
    "  # therefore group by exon (and deletion count), and compute \"sum deletion overlap\" in each group,\n",
    "  group_columns = df.columns.tolist()\n",
    "  group_columns.remove('deletion_overlap')\n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = group_columns,\n",
    "    aggregation_functions = {\n",
    "      'deletion_overlap': [np.sum],\n",
    "    }\n",
    "  )\n",
    "\n",
    "  df['exon_length'] = df['exon_end'] - df['exon_start']\n",
    "\n",
    "  # group exons by parent gene, \n",
    "  # computing \"sum deletion count\" and \"sum deletion overlap\" in each group, \n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = ['chromosome', 'gene_symbol'],\n",
    "    aggregation_functions = {\n",
    "      'exon_start': [np.min],\n",
    "      'exon_end': [np.max],\n",
    "      'exon_rank': 'count',\n",
    "      'deletion_count': [np.sum],\n",
    "      'sum deletion_overlap': [np.sum],\n",
    "      'exon_length': [np.sum],\n",
    "    }\n",
    "  )\n",
    "\n",
    "  df['gene coordinates'] = df['chromosome'] + ':' + df['min exon_start'].astype(str) + '-' + df['max exon_end'].astype(str)\n",
    "  df = df.drop(columns=['chromosome', 'min exon_start', 'max exon_end'])\n",
    "  df['deletion count per bp'] = df['sum deletion_count'] / df['sum exon_length']\n",
    "  df['fraction of bps that are deleted'] = df['sum sum deletion_overlap'] / df['sum exon_length']\n",
    "\n",
    "  df = df.rename(columns={\n",
    "    'count exon_rank': 'exon_count',\n",
    "    'sum deletion_count': 'deletion count over all exons',\n",
    "    'sum sum deletion_overlap': 'deletion overlap over all exons',\n",
    "    'sum exon_length': 'total length of all exons',\n",
    "  })\n",
    "\n",
    "  return df\n",
    "\n",
    "assign_SV_scores_to_labeled_genes(\n",
    "  gene_class='positive',\n",
    "  deletion_class='observed'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO \n",
    "# 1. assign a label to each enhancer\n",
    "# 2. compute PR curves, c.f., experiments/germline-model/chen-et-al-2022/SNV_plus_SV_model.3.ipynb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "constraint-tools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
