{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO: grand plan \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 1. Assign GC content, gBGC, BGS, exon status, enhancer status to Halldorsson windows using same infra I used for Chen windows \n",
    "# 2. Copy papers/neutral_models_are_biased/2.ipynb to new notebook, ingest Halldorsson dataframe, and compute PR curves (but not auPRC bar charts) for non-exonic Halldorsson windows \n",
    "# 3. Include PR curves for Halldorsson windows as supp info, showing that trends are same as when using Chen windows \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. intersect Halldorsson windows with enhancers and exons: \n",
    "#    papers/neutral_models_are_biased/6.Halldorsson/add-overlapAmounts.sh\n",
    "# 2. find noncoding Halldorsson windows and determine whether they significantly overlap enhancers or not: \n",
    "#    papers/neutral_models_are_biased/6.Halldorsson/assign_enhancer_and_exon_status.ipynb\n",
    "# 3. compute GC-content for Halldorsson windows: \n",
    "#    papers/neutral_models_are_biased/6.Halldorsson/compute-GC-content-for-all-window-sizes-based-on-Halldorsson-windows.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: delete 1, 2, 3, 4: \n",
    "#   1. assign BGS  experiments/germline-model/chen-et-al-2022/Chen_models_SNV_counts_best_for_most_frequent_windows.1.1.ipynb\n",
    "#   2. assign gBGC experiments/germline-model/chen-et-al-2022/Chen_models_SNV_counts_best_for_most_frequent_windows.1.2.ipynb\n",
    "#   3. assign GC content experiments/germline-model/chen-et-al-2022/Chen_models_SNV_counts_best_for_most_frequent_windows.7.ipynb\n",
    "#   4. also used: papers/neutral_models_are_biased/assign-depletion-rank-scores-to-chen-windows.ipynb\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assign BGS, gBGC to Halldorsson windows "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "CONSTRAINT_TOOLS = '/scratch/ucgd/lustre-labs/quinlan/u6018199/constraint-tools'\n",
    "CONSTRAINT_TOOLS_DATA = '/scratch/ucgd/lustre-labs/quinlan/data-shared/constraint-tools'\n",
    "\n",
    "import sys\n",
    "sys.path.append(f'{CONSTRAINT_TOOLS}/utilities')\n",
    "\n",
    "POP = 'EUR'\n",
    "GC_WINDOW_SIZE = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.intersect.B-map.bed\u001b[0m\n",
      "<ipython-input-6-81b67884980b>:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[col] = df[col].astype(dtype)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.intersect.paternal_recombination_rate-map.bed\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.intersect.maternal_recombination_rate-map.bed\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mWrote /scratch/ucgd/lustre-work/quinlan/data-shared/constraint-tools/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.maternal_recombination_rate.gBGC-tract-counts.bed\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gnocchi</th>\n",
       "      <th>N_expected</th>\n",
       "      <th>N_observed</th>\n",
       "      <th>window overlaps enhancer</th>\n",
       "      <th>window overlaps merged_exon</th>\n",
       "      <th>B</th>\n",
       "      <th>paternal_recombination_rate</th>\n",
       "      <th>maternal_recombination_rate</th>\n",
       "      <th>gBGC-tract_overlap_count</th>\n",
       "      <th>log10_paternal_recombination_rate</th>\n",
       "      <th>log10_maternal_recombination_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1432000</td>\n",
       "      <td>1433000</td>\n",
       "      <td>4.299894</td>\n",
       "      <td>338.059552</td>\n",
       "      <td>259</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.653</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1451000</td>\n",
       "      <td>1452000</td>\n",
       "      <td>0.666316</td>\n",
       "      <td>302.590657</td>\n",
       "      <td>291</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.652</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1453000</td>\n",
       "      <td>1454000</td>\n",
       "      <td>0.828398</td>\n",
       "      <td>269.601925</td>\n",
       "      <td>256</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.651</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1458000</td>\n",
       "      <td>1459000</td>\n",
       "      <td>-0.086128</td>\n",
       "      <td>270.583243</td>\n",
       "      <td>272</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.651</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1463000</td>\n",
       "      <td>1464000</td>\n",
       "      <td>2.948188</td>\n",
       "      <td>236.321790</td>\n",
       "      <td>191</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.651</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.025852</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.408422</td>\n",
       "      <td>-1.587499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556241</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137262000</td>\n",
       "      <td>137263000</td>\n",
       "      <td>-2.396149</td>\n",
       "      <td>263.131293</td>\n",
       "      <td>302</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.057053</td>\n",
       "      <td>0.063434</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.243724</td>\n",
       "      <td>-1.197675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556242</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137268000</td>\n",
       "      <td>137269000</td>\n",
       "      <td>3.640544</td>\n",
       "      <td>272.046527</td>\n",
       "      <td>212</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.057053</td>\n",
       "      <td>0.063434</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.243724</td>\n",
       "      <td>-1.197675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556243</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137269000</td>\n",
       "      <td>137270000</td>\n",
       "      <td>5.276351</td>\n",
       "      <td>300.458958</td>\n",
       "      <td>209</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.057053</td>\n",
       "      <td>0.063434</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.243724</td>\n",
       "      <td>-1.197675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556244</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137275000</td>\n",
       "      <td>137276000</td>\n",
       "      <td>2.687348</td>\n",
       "      <td>339.517045</td>\n",
       "      <td>290</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>1.564889</td>\n",
       "      <td>0.024254</td>\n",
       "      <td>0</td>\n",
       "      <td>0.194484</td>\n",
       "      <td>-1.615221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556245</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137290000</td>\n",
       "      <td>137291000</td>\n",
       "      <td>2.162811</td>\n",
       "      <td>257.721105</td>\n",
       "      <td>223</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.536</td>\n",
       "      <td>16.626887</td>\n",
       "      <td>0.384797</td>\n",
       "      <td>0</td>\n",
       "      <td>1.220811</td>\n",
       "      <td>-0.414769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1024014 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end   gnocchi  N_expected  N_observed  \\\n",
       "0        chr1    1432000    1433000  4.299894  338.059552         259   \n",
       "4        chr1    1451000    1452000  0.666316  302.590657         291   \n",
       "5        chr1    1453000    1454000  0.828398  269.601925         256   \n",
       "7        chr1    1458000    1459000 -0.086128  270.583243         272   \n",
       "11       chr1    1463000    1464000  2.948188  236.321790         191   \n",
       "...       ...        ...        ...       ...         ...         ...   \n",
       "1556241  chr9  137262000  137263000 -2.396149  263.131293         302   \n",
       "1556242  chr9  137268000  137269000  3.640544  272.046527         212   \n",
       "1556243  chr9  137269000  137270000  5.276351  300.458958         209   \n",
       "1556244  chr9  137275000  137276000  2.687348  339.517045         290   \n",
       "1556245  chr9  137290000  137291000  2.162811  257.721105         223   \n",
       "\n",
       "         window overlaps enhancer  window overlaps merged_exon      B  \\\n",
       "0                            True                        False  0.653   \n",
       "4                           False                        False  0.652   \n",
       "5                           False                        False  0.651   \n",
       "7                           False                        False  0.651   \n",
       "11                           True                        False  0.651   \n",
       "...                           ...                          ...    ...   \n",
       "1556241                      True                        False  0.536   \n",
       "1556242                      True                        False  0.536   \n",
       "1556243                      True                        False  0.536   \n",
       "1556244                      True                        False  0.536   \n",
       "1556245                     False                        False  0.536   \n",
       "\n",
       "         paternal_recombination_rate  maternal_recombination_rate  \\\n",
       "0                           0.039046                     0.025852   \n",
       "4                           0.039046                     0.025852   \n",
       "5                           0.039046                     0.025852   \n",
       "7                           0.039046                     0.025852   \n",
       "11                          0.039046                     0.025852   \n",
       "...                              ...                          ...   \n",
       "1556241                     0.057053                     0.063434   \n",
       "1556242                     0.057053                     0.063434   \n",
       "1556243                     0.057053                     0.063434   \n",
       "1556244                     1.564889                     0.024254   \n",
       "1556245                    16.626887                     0.384797   \n",
       "\n",
       "         gBGC-tract_overlap_count  log10_paternal_recombination_rate  \\\n",
       "0                               0                          -1.408422   \n",
       "4                               0                          -1.408422   \n",
       "5                               0                          -1.408422   \n",
       "7                               0                          -1.408422   \n",
       "11                              0                          -1.408422   \n",
       "...                           ...                                ...   \n",
       "1556241                         0                          -1.243724   \n",
       "1556242                         0                          -1.243724   \n",
       "1556243                         0                          -1.243724   \n",
       "1556244                         0                           0.194484   \n",
       "1556245                         0                           1.220811   \n",
       "\n",
       "         log10_maternal_recombination_rate  \n",
       "0                                -1.587499  \n",
       "4                                -1.587499  \n",
       "5                                -1.587499  \n",
       "7                                -1.587499  \n",
       "11                               -1.587499  \n",
       "...                                    ...  \n",
       "1556241                          -1.197675  \n",
       "1556242                          -1.197675  \n",
       "1556243                          -1.197675  \n",
       "1556244                          -1.615221  \n",
       "1556245                          -0.414769  \n",
       "\n",
       "[1024014 rows x 14 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import polars as pl # TODO \n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "\n",
    "from shell import shell \n",
    "from colorize import print_string_as_info\n",
    "from aggregate import aggregate\n",
    "\n",
    "def get_header(filename):\n",
    "  with open(filename) as fh: \n",
    "    header = fh.readline().strip().split('\\t')\n",
    "  return header\n",
    "\n",
    "def intersect(a_filename, b_filename, intersect_filename, b_class): \n",
    "  cmd = (\n",
    "    f'bedtools intersect'\n",
    "    f\" -a <(tail -n +2 {a_filename})\" # contains header\n",
    "    f\" -b <(tail -n +2 {b_filename})\" # contains header\n",
    "    f' -wao -f 0.5 -r' \n",
    "    f' > {intersect_filename}'\n",
    "  )\n",
    "  shell(cmd)\n",
    "  print_string_as_info(f'Wrote {intersect_filename}')\n",
    "\n",
    "  a_file_header = get_header(a_filename) # contains header\n",
    "  b_file_header = get_header(b_filename) # contains header\n",
    "  b_file_header = [f'{b_class}_{field}' for field in b_file_header[:3]] + b_file_header[3:]\n",
    "\n",
    "  return a_file_header, b_file_header\n",
    "\n",
    "def set_column_dtypes(df, column_dtypes, log=False):\n",
    "  df = df.replace('.', np.nan)\n",
    "\n",
    "  for col, dtype in column_dtypes.items():\n",
    "    df[col] = df[col].astype(dtype)\n",
    "\n",
    "  if log: print(dict(df.dtypes))\n",
    "  return df \n",
    "\n",
    "def make_scores_numeric(df, b_class, b_features): \n",
    "  # we don't want to include windows that don't overlap any windows with b_feature values\n",
    "  df = df[df[f'{b_class}_chromosome'] != '.'] \n",
    "\n",
    "  # having removed records where 'b_class_chromosome' is '.', we can now convert the 'b_features' columns to floats\n",
    "  column_dtypes = {b_feature: float for b_feature in b_features}\n",
    "  df = set_column_dtypes(df, column_dtypes)\n",
    "\n",
    "  return df \n",
    "  \n",
    "def intersect_and_aggregate(a_filename_stem, b_filename, b_class, b_features, b_class_aggregation): \n",
    "  a_filename = f'{a_filename_stem}.bed'\n",
    "  intersect_filename = f'{a_filename_stem}.intersect.{b_class}.bed'\n",
    "\n",
    "  a_file_header, b_file_header = intersect(a_filename, b_filename, intersect_filename, b_class)\n",
    "\n",
    "  df = pd.read_csv(\n",
    "    intersect_filename, \n",
    "    sep='\\t',\n",
    "    names=a_file_header + b_file_header + [f'halldorssonWindow_{b_class}Window_overlap'],\n",
    "  )\n",
    "\n",
    "  df = make_scores_numeric(df, b_class, b_features)\n",
    "\n",
    "  # some Halldorsson-windows may intersect multiple b_feature-windows, \n",
    "  # so let's group by Halldorsson-window, and aggregate scores over all b_feature-windows in the group\n",
    "  df = aggregate(\n",
    "    df, \n",
    "    group_columns = a_file_header,\n",
    "    aggregation_functions = {b_feature: [b_class_aggregation] for b_feature in b_features}\n",
    "  )\n",
    "  df = df.rename(columns={f'{b_class_aggregation} {b_feature}': b_feature for b_feature in b_features})\n",
    "\n",
    "  # TODO: \n",
    "  df = df.rename(columns={'rank': b_class})\n",
    "  \n",
    "  df.to_csv(\n",
    "    f'{a_filename_stem}.{b_class}.bed',\n",
    "    sep='\\t',\n",
    "    index=False,\n",
    "  )\n",
    "\n",
    "  print_string_as_info(f'Wrote {a_filename_stem}.{b_class}.bed')\n",
    "\n",
    "  return df \n",
    "\n",
    "# TODO: continue here: \n",
    "def assign_BGS_and_gBGC_to_halldorsson_windows():\n",
    "  # assign BGS values to Halldorsson windows \n",
    "  intersect_and_aggregate(\n",
    "    a_filename_stem = f'{CONSTRAINT_TOOLS_DATA}/depletion_rank_scores/41586_2022_4965_MOESM3_ESM.noncoding.enhancer', \n",
    "    b_filename = f'{CONSTRAINT_TOOLS_DATA}/background-selection/CADD-B-map/bmap.hg38.header.bed', \n",
    "    b_class = 'XXX',\n",
    "    b_features = ['YYY'], \n",
    "    b_class_aggregation = 'min' \n",
    "  )   \n",
    "\n",
    "  return \n",
    "\n",
    "  # assign gBGC coefficients to Halldorsson windows \n",
    "  df = intersect_and_aggregate(\n",
    "    a_filename_stem = f'{CONSTRAINT_TOOLS_DATA}/depletion_rank_scores/41586_2022_4965_MOESM3_ESM.noncoding.enhancer.B', \n",
    "    b_filename = f'{CONSTRAINT_TOOLS_DATA}/GC-biased-gene-conversion/gBGC-coefficient.hg38.{POP}.bed', # TODO: does this bed file have a header, as the code assumes? \n",
    "    b_class = 'XXX',\n",
    "    b_features = ['YYY'], \n",
    "    b_class_aggregation = 'mean'\n",
    "  )   \n",
    "\n",
    "  return df\n",
    "\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "WINDOWS = assign_BGS_and_gBGC_to_halldorsson_windows()\n",
    "WINDOWS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO: Assign GC content to Halldorsson windows "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from functools import reduce\n",
    "\n",
    "def get_windows_with_GC_content_and_cpg_islands(): \n",
    "  df1 = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.gnocchi.N_expected.N_observed.B.paternal_recombination_rate.maternal_recombination_rate.gBGC-tract-counts.non-exonic.gBGC.bed', \n",
    "    sep='\\t', \n",
    "  )\n",
    "\n",
    "  df2 = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/chen-et-al-2023-published-version/41586_2023_6045_MOESM4_ESM/Supplementary_Data_2.GC_content_{GC_WINDOW_SIZE}.bed', \n",
    "    sep='\\t', \n",
    "  )\n",
    "  df2 = df2[['chen_chrom', 'chen_start', 'chen_end', 'window_GC_content']]\n",
    "  df2 = df2.rename(columns={\n",
    "    'chen_chrom': 'chrom', \n",
    "    'chen_start': 'start', \n",
    "    'chen_end': 'end',\n",
    "    'window_GC_content': f'GC_content_{GC_WINDOW_SIZE}bp'\n",
    "  })\n",
    "\n",
    "  # created using: experiments/germline-model/chen-et-al-2022/cpg-island-enrichment.ipynb\n",
    "  df3 = pd.read_csv(\n",
    "    f'{CONSTRAINT_TOOLS_DATA}/benchmark-genome-wide-predictions/chen-et-al-2022/mchale.kmerSizes.trainSets.noisy.enhancer-exon-cpgIsland.bed',\n",
    "    sep='\\t', \n",
    "  )\n",
    "  df3 = df3[['chromosome', 'start', 'end', 'cpg_island overlap', 'window overlaps cpg_island']]\n",
    "  df3 = df3.rename(columns={\n",
    "    'chromosome': 'chrom', \n",
    "    'cpg_island overlap': 'cpg_island_overlap', \n",
    "    'window overlaps cpg_island': 'window_overlaps_cpg_island'\n",
    "  })\n",
    "\n",
    "  dfs = [df1, df2, df3]\n",
    "  df = reduce(lambda left, right: pd.merge(left, right, on=['chrom', 'start', 'end'], how='inner'), dfs)\n",
    "\n",
    "  df = df[df['window overlaps enhancer'] == False]\n",
    "\n",
    "  return df\n",
    "\n",
    "WINDOWS = get_windows_with_GC_content_and_cpg_islands()\n",
    "WINDOWS"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "constraint-tools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
